--- a/loginutils/passwd.c
+++ b/loginutils/passwd.c
@@ -40,6 +40,34 @@
 #include "libbb.h"
 #include <syslog.h>
 
+#ifdef CONFIG_PASSWD
+#include <liblog.h>
+
+void log_event(char *user) {
+	struct events_log new_task = {0};
+	struct ubus_context *ubus;
+	char cmd[BUFFER_SIZE*8];
+
+	ubus = ubus_connect(NULL);
+	if (!ubus){
+		syslog(LOG_ERR, "Failed to connect to ubus");
+		return;
+	}
+
+	cmd[0] = '\0';
+	sprintf(cmd, "password has been changed for user %s", user);
+
+	new_task.table = LLOG_CONNECTIONS;
+	new_task.sender = "Passwd";
+	new_task.priority = EVENTLOG_NOTICE;
+	new_task.db_flag = 1;
+	new_task.text = cmd;
+
+	llog_add_task(ubus, &new_task);
+	ubus_free(ubus);
+}
+#endif // CONFIG_PASSWD
+
 static char* new_password(const struct passwd *pw, uid_t myuid, const char *algo)
 {
 	char salt[MAX_PW_SALT_LEN];
@@ -232,7 +260,11 @@ int passwd_main(int argc UNUSED_PARAM, c
 	/* LOGMODE_BOTH */
 	if (rc < 0)
 		bb_error_msg_and_die("can't update password file %s", filename);
+#ifndef CONFIG_PASSWD
 	bb_info_msg("password for %s changed by %s", name, myname);
+#else
+	log_event(name);
+#endif // CONFIG_PASSWD
 
 	/*if (ENABLE_FEATURE_CLEAN_UP) free(newp); - can't, it may be non-malloced */
  skip:
--- a/Makefile
+++ b/Makefile
@@ -572,6 +572,10 @@ libs-y1		:= $(patsubst %/, %/lib.a, $(li
 libs-y2		:= $(patsubst %/, %/built-in.o, $(libs-y))
 libs-y		:= $(libs-y1) $(libs-y2)
 
+ifdef CONFIG_PASSWD
+CONFIG_EXTRA_LDLIBS:=ubus log
+endif
+
 # Build busybox
 # ---------------------------------------------------------------------------
 # busybox is build from the objects selected by $(busybox-init) and
