diff --git a/loginutils/getty.c b/loginutils/getty.c
index 6c6d409..7a6846d 100644
--- a/loginutils/getty.c
+++ b/loginutils/getty.c
@@ -221,8 +221,8 @@ static void open_tty(void)
 		xopen(G.tty_name, O_RDWR | O_NONBLOCK); /* uses fd 0 */
 
 		/* Set proper protections and ownership */
-		fchown(0, 0, 0);        /* 0:0 */
-		fchmod(0, 0620);        /* crw--w---- */
+//		fchown(0, 0, 0);        /* 0:0 */
+//		fchmod(0, 0620);        /* crw--w---- */
 	} else {
 		char *n;
 		/*
