diff --git a/loginutils/login.c b/loginutils/login.c
index ce87e31..85c82a9 100644
--- a/loginutils/login.c
+++ b/loginutils/login.c
@@ -555,8 +555,8 @@ int login_main(int argc UNUSED_PARAM, char **argv)
 
 	/* Try these, but don't complain if they fail.
 	 * _f_chown is safe wrt race t=ttyname(0);...;chown(t); */
-	fchown(0, pw->pw_uid, pw->pw_gid);
-	fchmod(0, 0600);
+//	fchown(0, pw->pw_uid, pw->pw_gid);
+//	fchmod(0, 0600);
 
 	/* We trust environment only if we run by root */
 	if (ENABLE_LOGIN_SCRIPTS && run_by_root)
