--- a/net/netfilter/nf_conntrack_standalone.c
+++ b/net/netfilter/nf_conntrack_standalone.c
@@ -522,18 +522,18 @@ static int nf_conntrack_standalone_init_
 {
 	struct proc_dir_entry *pde;
 	kuid_t root_uid;
-	kgid_t root_gid;
+	kgid_t network_gid;
 
-	pde = proc_create_net_data_write("nf_conntrack", 0440, net->proc_net,
+	pde = proc_create_net_data_write("nf_conntrack", 0460, net->proc_net,
 					 &ct_seq_ops, &ct_file_write,
 					 sizeof(struct ct_iter_state), NULL);
 	if (!pde)
 		goto out_nf_conntrack;
 
 	root_uid = make_kuid(net->user_ns, 0);
-	root_gid = make_kgid(net->user_ns, 0);
-	if (uid_valid(root_uid) && gid_valid(root_gid))
-		proc_set_user(pde, root_uid, root_gid);
+	network_gid = make_kgid(net->user_ns, 101);
+	if (uid_valid(root_uid) && gid_valid(network_gid))
+		proc_set_user(pde, root_uid, network_gid);
 
 	pde = proc_create_net("nf_conntrack", 0444, net->proc_net_stat,
 			&ct_cpu_seq_ops, sizeof(struct seq_net_private));
--- a/net/netfilter/x_tables.c
+++ b/net/netfilter/x_tables.c
@@ -1817,6 +1817,7 @@ int xt_proto_init(struct net *net, u_int
 	struct proc_dir_entry *proc;
 	kuid_t root_uid;
 	kgid_t root_gid;
+	kgid_t network_gid;
 #endif
 
 	if (af >= ARRAY_SIZE(xt_prefix))
@@ -1826,16 +1827,17 @@ int xt_proto_init(struct net *net, u_int
 #ifdef CONFIG_PROC_FS
 	root_uid = make_kuid(net->user_ns, 0);
 	root_gid = make_kgid(net->user_ns, 0);
+	network_gid = make_kgid(net->user_ns, 101);
 
 	strscpy(buf, xt_prefix[af], sizeof(buf));
 	strlcat(buf, FORMAT_TABLES, sizeof(buf));
-	proc = proc_create_net_data(buf, 0440, net->proc_net, &xt_table_seq_ops,
+	proc = proc_create_net_data(buf, 0460, net->proc_net, &xt_table_seq_ops,
 			sizeof(struct seq_net_private),
 			(void *)(unsigned long)af);
 	if (!proc)
 		goto out;
-	if (uid_valid(root_uid) && gid_valid(root_gid))
-		proc_set_user(proc, root_uid, root_gid);
+	if (uid_valid(root_uid) && gid_valid(network_gid))
+		proc_set_user(proc, root_uid, network_gid);
 
 	strscpy(buf, xt_prefix[af], sizeof(buf));
 	strlcat(buf, FORMAT_MATCHES, sizeof(buf));
