--- a/include/uapi/linux/ethtool.h
+++ b/include/uapi/linux/ethtool.h
@@ -288,12 +288,18 @@ struct ethtool_tunable {
 #define ETHTOOL_PHY_EDPD_DFLT_TX_MSECS		0xffff
 #define ETHTOOL_PHY_EDPD_NO_TX			0xfffe
 #define ETHTOOL_PHY_EDPD_DISABLE		0
+#define ETHTOOL_PHY_LED_MODE_OFF		0
+#define ETHTOOL_PHY_LED_MODE_ON			1
+#define ETHTOOL_PHY_LED_MODE_AUTO		2
+#define ETHTOOL_PHY_LED_MODE_UNKNOWN		3
+
 
 enum phy_tunable_id {
 	ETHTOOL_PHY_ID_UNSPEC,
 	ETHTOOL_PHY_DOWNSHIFT,
 	ETHTOOL_PHY_FAST_LINK_DOWN,
 	ETHTOOL_PHY_EDPD,
+	ETHTOOL_PHY_LED_MODE,
 	/*
 	 * Add your fresh new phy tunable attribute above and remember to update
 	 * phy_tunable_strings[] in net/ethtool/common.c
--- a/net/ethtool/common.c
+++ b/net/ethtool/common.c
@@ -98,6 +98,7 @@ phy_tunable_strings[__ETHTOOL_PHY_TUNABL
 	[ETHTOOL_PHY_DOWNSHIFT]	= "phy-downshift",
 	[ETHTOOL_PHY_FAST_LINK_DOWN] = "phy-fast-link-down",
 	[ETHTOOL_PHY_EDPD]	= "phy-energy-detect-power-down",
+	[ETHTOOL_PHY_LED_MODE]	= "phy-led-mode",
 };
 
 #define __LINK_MODE_NAME(speed, type, duplex) \
--- a/net/ethtool/ioctl.c
+++ b/net/ethtool/ioctl.c
@@ -2655,6 +2655,11 @@ static int ethtool_phy_tunable_valid(con
 		    tuna->type_id != ETHTOOL_TUNABLE_U16)
 			return -EINVAL;
 		break;
+	case ETHTOOL_PHY_LED_MODE:
+		if (tuna->len != sizeof(u8) ||
+		    tuna->type_id != ETHTOOL_TUNABLE_U8)
+		    	return -EINVAL;
+		break;
 	default:
 		return -EINVAL;
 	}
