--- a/drivers/net/usb/usbnet.c
+++ b/drivers/net/usb/usbnet.c
@@ -30,6 +30,7 @@
 #include <linux/slab.h>
 #include <linux/kernel.h>
 #include <linux/pm_runtime.h>
+#include <linux/if_arp.h>
 
 /*-------------------------------------------------------------------------*/
 
@@ -404,7 +405,7 @@ int usbnet_change_mtu (struct net_device
 	int		old_rx_urb_size = dev->rx_urb_size;
 
 	// no second zero-length packet read wanted after mtu-sized packets
-	if ((ll_mtu % dev->maxpacket) == 0)
+	if (((ll_mtu % dev->maxpacket) == 0) && (net->type != ARPHRD_NONE))
 		return -EDOM;
 	net->mtu = new_mtu;
 
