--- a/kernel/bpf/syscall.c
+++ b/kernel/bpf/syscall.c
@@ -893,6 +893,11 @@ static int map_create(union bpf_attr *at
 			if (err)
 				goto free_map;
 		}
+		if (btf_is_kernel(btf)) {
+			btf_put(btf);
+			err = -EACCES;
+			goto free_map;
+		}
 
 		map->btf_key_type_id = attr->btf_key_type_id;
 		map->btf_value_type_id = attr->btf_value_type_id;
--- a/kernel/bpf/verifier.c
+++ b/kernel/bpf/verifier.c
@@ -9196,6 +9196,11 @@ static int check_btf_info(struct bpf_ver
 	btf = btf_get_by_fd(attr->prog_btf_fd);
 	if (IS_ERR(btf))
 		return PTR_ERR(btf);
+	if (btf_is_kernel(btf)) {
+		btf_put(btf);
+		return -EACCES;
+	}
+
 	env->prog->aux->btf = btf;
 
 	err = check_btf_func(env, attr, uattr);
--- a/include/linux/btf.h
+++ b/include/linux/btf.h
@@ -89,6 +89,7 @@ int btf_type_snprintf_show(const struct
 
 int btf_get_fd_by_id(u32 id);
 u32 btf_id(const struct btf *btf);
+bool btf_is_kernel(const struct btf *btf);
 bool btf_member_is_reg_int(const struct btf *btf, const struct btf_type *s,
 			   const struct btf_member *m,
 			   u32 expected_offset, u32 expected_size);
--- a/kernel/bpf/btf.c
+++ b/kernel/bpf/btf.c
@@ -210,6 +210,7 @@ struct btf {
 	refcount_t refcnt;
 	u32 id;
 	struct rcu_head rcu;
+	bool kernel_btf;
 };
 
 enum verifier_phase {
@@ -430,6 +431,11 @@ static bool btf_type_nosize(const struct
 	       btf_type_is_func(t) || btf_type_is_func_proto(t);
 }
 
+bool btf_is_kernel(const struct btf *btf)
+{
+	return btf->kernel_btf;
+}
+
 static bool btf_type_nosize_or_null(const struct btf_type *t)
 {
 	return !t || btf_type_nosize(t);
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -4424,6 +4424,7 @@ struct bpf_btf_info {
 	__aligned_u64 btf;
 	__u32 btf_size;
 	__u32 id;
+	__u32 kernel_btf;
 } __attribute__((aligned(8)));
 
 struct bpf_link_info {
