Index: linux-5.10.199/net/netfilter/nf_tables_api.c
===================================================================
--- linux-5.10.199.orig/net/netfilter/nf_tables_api.c
+++ linux-5.10.199/net/netfilter/nf_tables_api.c
@@ -8360,7 +8360,7 @@ static void nft_set_commit_update(struct
 	list_for_each_entry_safe(set, next, set_update_list, pending_update) {
 		list_del_init(&set->pending_update);
 
-		if (!set->ops->commit)
+		if (!set->ops->commit || set->dead)
 			continue;
 
 		set->ops->commit(set);
