From 2f8b1360df25bab375ec60bbba2dce8dd796161c Mon Sep 17 00:00:00 2001
From: Jo-Philipp Wich <jo@mein.io>
Date: Mon, 7 Feb 2022 10:04:11 +0100
Subject: [PATCH] main: fix leaking -p/-s argument values

Signed-off-by: Jo-Philipp Wich <jo@mein.io>
---
 main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/main.c b/main.c
index d1aac09..de5f2c3 100644
--- a/main.c
+++ b/main.c
@@ -328,6 +328,7 @@ int main(int argc, char **argv)
 		case 'p':
 			optarg = strdup(optarg);
 			bound += add_listener_arg(optarg, (ch == 's'));
+			free(optarg);
 			break;
 
 		case 'h':
-- 
2.30.2