Index: uhttpd-2021-03-21-15346de8/proc.c
===================================================================
--- uhttpd-2021-03-21-15346de8.orig/proc.c
+++ uhttpd-2021-03-21-15346de8/proc.c
@@ -36,6 +36,8 @@
 	__header(content_type, content-type) \
 	__header(content_length, content-length) \
 	__header(x_http_method_override, x-http-method-override) \
+	__header(x_csrf_protection, x-csrf-protection) \
+	__header(x_last_event_id, x-last-event-id) \
 	__header(http_auth_user, http-auth-user) \
 	__header(http_auth_pass, http-auth-pass)
 
@@ -68,6 +70,8 @@ static const struct {
 	{ "HTTP_REFERER", HDR_referer },
 	{ "HTTP_USER_AGENT", HDR_user_agent },
 	{ "HTTP_X_HTTP_METHOD_OVERRIDE", HDR_x_http_method_override },
+	{ "HTTP_X_CSRF_PROTECTION", HDR_x_csrf_protection },
+	{ "HTTP_X_LAST_EVENT_ID", HDR_x_last_event_id },
 	{ "HTTP_AUTH_USER", HDR_http_auth_user },
 	{ "HTTP_AUTH_PASS", HDR_http_auth_pass },
 	{ "CONTENT_TYPE", HDR_content_type },
