--- a/client.c
+++ b/client.c
@@ -69,8 +69,8 @@ static void add_security_headers(struct
 		}
 		ustream_printf(cl->us, "Content-Security-Policy: \
 				default-src 'self'; frame-src * 'self'; \
-				img-src * 'self' data:; \
-				script-src 'wasm-unsafe-eval' https://cdn.jsdelivr.net/ 'sha256-LV+4Blj3LIWsLHvvq37tdHwBQpq6ZOI+wHw3XeGsJzM=' 'self' 'sha256-8yE2w7Bv8/Il8SvtEkB35j3QNRei9CKtdX8HgqblU04='%s 'sha256-Pv4HyWUIdh/mQalp8JMzRdM1eTkfzYhRRnW/9m4pzQ8=' blob:; \
+				img-src * 'self' data: blob:; \
+				script-src 'wasm-unsafe-eval' https://cdn.jsdelivr.net/ 'sha256-LV+4Blj3LIWsLHvvq37tdHwBQpq6ZOI+wHw3XeGsJzM=' 'self' 'sha256-8yE2w7Bv8/Il8SvtEkB35j3QNRei9CKtdX8HgqblU04='%s 'sha256-Pv4HyWUIdh/mQalp8JMzRdM1eTkfzYhRRnW/9m4pzQ8=' 'sha256-/83V68bpILlL4sGTUJ9oviiw3La/Bskmmf83WaWOw9I=' blob:; \
 				style-src https://cdn.jsdelivr.net/ 'unsafe-hashes' 'sha256-4Su6mBWzEIFnH4pAGMOuaeBrstwJN4Z3pq/s1Kn4/KQ=' 'sha256-kwpt3lQZ21rs4cld7/uEm9qI5yAbjYzx+9FGm/XmwNU=' 'self'; \
 				connect-src 'self'; \
 				worker-src 'self' blob:; \
