Index: 7zip-22.0.1/CPP/7zip/UI/Console/Main.cpp
===================================================================
--- 7zip-22.0.1.orig/CPP/7zip/UI/Console/Main.cpp
+++ 7zip-22.0.1/CPP/7zip/UI/Console/Main.cpp
@@ -98,8 +98,7 @@ static const char * const kCopyrightStri
   " : " MY_COPYRIGHT_DATE "\n";
 
 static const char * const kHelpString =
-    "Usage: 7z"
-    PROG_POSTFIX
+    "Usage: 7zr"
     " <command> [<switches>...] <archive_name> [<file_names>...] [@listfile]\n"
     "\n"
     "<Commands>\n"
