Index: backports-6.1.24/drivers/net/wireless/ath/ath10k/mac.c
===================================================================
--- backports-6.1.24.orig/drivers/net/wireless/ath/ath10k/mac.c	2024-03-12 13:57:45.183025499 +0200
+++ backports-6.1.24/drivers/net/wireless/ath/ath10k/mac.c	2024-03-12 14:01:28.441545575 +0200
@@ -4543,16 +4545,15 @@
 	case ATH10K_SCAN_ABORTING:
 		if (!ar->scan.is_roc) {
 			struct cfg80211_scan_info info = {
-				.aborted = (ar->scan.state ==
-					    ATH10K_SCAN_ABORTING),
+				.aborted = ((ar->scan.state ==
+					    ATH10K_SCAN_ABORTING) ||
+					    (ar->scan.state ==
+					    ATH10K_SCAN_STARTING)),
 			};
 
 			ieee80211_scan_completed(ar->hw, &info);
-		} else if (ar->scan.roc_notify) {
-			ieee80211_remain_on_channel_expired(ar->hw);
 		}
-		fallthrough;
-	case ATH10K_SCAN_STARTING:
+
 		ar->scan.state = ATH10K_SCAN_IDLE;
 		ar->scan_channel = NULL;
 		ar->scan.roc_freq = 0;
