Index: backports-6.1.24/drivers/net/wireless/ath/ath10k/htt_tx.c
===================================================================
--- backports-6.1.24.orig/drivers/net/wireless/ath/ath10k/htt_tx.c	2023-04-17 17:56:40.000000000 +0300
+++ backports-6.1.24/drivers/net/wireless/ath/ath10k/htt_tx.c	2024-06-18 08:58:19.817160426 +0300
@@ -1423,16 +1423,22 @@
 
 	if (!is_eth) {
 		struct ieee80211_hdr *hdr = (struct ieee80211_hdr *)msdu->data;
+		unsigned int mic_len = IEEE80211_CCMP_MIC_LEN;
+
+		if (skb_cb->ucast_cipher == WLAN_CIPHER_SUITE_GCMP ||
+			skb_cb->ucast_cipher == WLAN_CIPHER_SUITE_GCMP_256) {
+			mic_len = IEEE80211_GCMP_MIC_LEN;
+		}
 
 		if ((ieee80211_is_action(hdr->frame_control) ||
 		     ieee80211_is_deauth(hdr->frame_control) ||
 		     ieee80211_is_disassoc(hdr->frame_control)) &&
 		     ieee80211_has_protected(hdr->frame_control)) {
-			skb_put(msdu, IEEE80211_CCMP_MIC_LEN);
+			skb_put(msdu, mic_len);
 		} else if (!(skb_cb->flags & ATH10K_SKB_F_NO_HWCRYPT) &&
 			   txmode == ATH10K_HW_TXRX_RAW &&
 			   ieee80211_has_protected(hdr->frame_control)) {
-			skb_put(msdu, IEEE80211_CCMP_MIC_LEN);
+			skb_put(msdu, mic_len);
 		}
 	}
 
