Index: netifd-2024-01-04-c18cc79d/scripts/netifd-wireless.sh
===================================================================
--- netifd-2024-01-04-c18cc79d.orig/scripts/netifd-wireless.sh
+++ netifd-2024-01-04-c18cc79d/scripts/netifd-wireless.sh
@@ -218,6 +218,13 @@ wireless_vif_parse_encryption() {
 		wpa_cipher="CCMP"
 	fi
 
+	# WPA3 enterprise requires the GCMP-256 cipher (technically also CCMP and GCMP are possible
+	# but many clients/devices do not support that)
+	case "$encryption" in
+		wpa3-mixed*) wpa_cipher="${wpa_cipher} GCMP-256";;
+		wpa3*) wpa_cipher="GCMP-256";;
+	esac
+
 	case "$encryption" in
 		*tkip+aes|*tkip+ccmp|*aes+tkip|*ccmp+tkip) wpa_cipher="CCMP TKIP";;
 		*ccmp256) wpa_cipher="CCMP-256";;
