Index: netifd-2024-01-04-c18cc79d/config.c
===================================================================
--- netifd-2024-01-04-c18cc79d.orig/config.c
+++ netifd-2024-01-04-c18cc79d/config.c
@@ -673,8 +673,26 @@ config_init_wireless(void)
 			continue;
 
 		dev_name = uci_lookup_option_string(uci_ctx, s, "device");
-		if (!dev_name)
-			continue;
+		if (!dev_name) {
+			struct uci_option *opt = uci_lookup_option(uci_ctx, s, "device");
+			struct uci_element *el;
+
+			if(!opt || opt->type != UCI_TYPE_LIST) {
+				continue;
+			}
+
+			uci_foreach_element(&opt->v.list, el) {
+				dev_name = el->name;
+				wdev = vlist_find(&wireless_devices, dev_name, wdev, node);
+				if (!wdev) {
+					DPRINTF("device %s not found!\n", dev_name);
+					continue;
+				}
+
+				config_parse_wireless_interface(wdev, s);
+			}
+ 			continue;
+		}
 
 		wdev = vlist_find(&wireless_devices, dev_name, wdev, node);
 		if (!wdev) {
