Index: netifd-2024-01-04-c18cc79d/scripts/netifd-wireless.sh
===================================================================
--- netifd-2024-01-04-c18cc79d.orig/scripts/netifd-wireless.sh	2024-02-21 12:37:04.218351512 +0200
+++ netifd-2024-01-04-c18cc79d/scripts/netifd-wireless.sh	2024-02-21 12:38:07.153988931 +0200
@@ -218,6 +218,13 @@
 		wpa_cipher="CCMP"
 	fi
 
+	# WPA3 enterprise requires the GCMP-256 cipher (technically also CCMP and GCMP are possible
+	# but many clients/devices do not support that)
+	case "$encryption" in
+		wpa3-mixed*) wpa_cipher="${wpa_cipher} GCMP-256";;
+		wpa3*) wpa_cipher="GCMP-256";;
+	esac
+
 	case "$encryption" in
 		*tkip+aes|*tkip+ccmp|*aes+tkip|*ccmp+tkip) wpa_cipher="CCMP TKIP";;
 		*ccmp256) wpa_cipher="CCMP-256";;
@@ -225,7 +232,6 @@
 		*tkip) wpa_cipher="TKIP";;
 		*gcmp256) wpa_cipher="GCMP-256";;
 		*gcmp) wpa_cipher="GCMP";;
-		wpa3-192*) wpa_cipher="GCMP-256";;
 	esac
 
 	# 802.11n requires CCMP for WPA
@@ -257,14 +263,11 @@
 		owe*)
 			auth_type=owe
 		;;
-		wpa3-192*)
-			auth_type=eap192
-		;;
 		wpa3-mixed*)
-			auth_type=eap-eap2
+			auth_type=eap-eap192
 		;;
 		wpa3*)
-			auth_type=eap2
+			auth_type=eap192
 		;;
 		psk3-mixed*|sae-mixed*)
 			auth_type=psk-sae
