Index: netifd-2024-01-04-c18cc79d/wireless.c
===================================================================
--- netifd-2024-01-04-c18cc79d.orig/wireless.c
+++ netifd-2024-01-04-c18cc79d/wireless.c
@@ -661,6 +661,9 @@ wireless_device_mark_up(struct wireless_
 	netifd_log_message(L_NOTICE, "Wireless device '%s' is now up\n", wdev->name);
 	wdev->retry = WIRELESS_SETUP_RETRY;
 	wdev->state = IFS_UP;
+	free(wdev->prev_config);
+	wdev->prev_config = NULL;
+
 	vlist_for_each_element(&wdev->interfaces, vif, node) {
 		wireless_interface_handle_link(vif, NULL, true);
 		vlist_for_each_element(&vif->vlans, vlan, node)
