Index: netifd-2022-01-12-5ca5e0b4/device.c
===================================================================
--- netifd-2022-01-12-5ca5e0b4.orig/device.c
+++ netifd-2022-01-12-5ca5e0b4/device.c
@@ -1143,20 +1143,21 @@ device_dump_status(struct blob_buf *b, s
 	blobmsg_add_u8(b, "carrier", !!dev->link_active);
 	blobmsg_add_u8(b, "auth_status", !!dev->auth_status);
 
+	system_if_get_settings(dev, &dev->settings);
+	device_merge_settings(dev, &st);
+	if (st.flags & DEV_OPT_MACADDR)
+		blobmsg_add_string(b, "macaddr", format_macaddr(st.macaddr));
+
 	if (dev->type->dump_info)
 		dev->type->dump_info(dev, b);
 	else
 		system_if_dump_info(dev, b);
 
 	if (dev->active) {
-		system_if_get_settings(dev, &dev->settings);
-		device_merge_settings(dev, &st);
 		if (st.flags & DEV_OPT_MTU)
 			blobmsg_add_u32(b, "mtu", st.mtu);
 		if (st.flags & DEV_OPT_MTU6)
 			blobmsg_add_u32(b, "mtu6", st.mtu6);
-		if (st.flags & DEV_OPT_MACADDR)
-			blobmsg_add_string(b, "macaddr", format_macaddr(st.macaddr));
 		if (st.flags & DEV_OPT_TXQUEUELEN)
 			blobmsg_add_u32(b, "txqueuelen", st.txqueuelen);
 		if (st.flags & DEV_OPT_IPV6)
