--- a/system-linux.c
+++ b/system-linux.c
@@ -49,6 +49,8 @@
 
 #include <sched.h>
 
+#include <linux/version.h>
+
 #include "ethtool-modes.h"
 
 #ifndef RTN_FAILED_POLICY
@@ -3841,6 +3843,7 @@ static int system_iprule(struct iprule *
 	if (rule->flags & IPRULE_SUP_PREFIXLEN)
 		nla_put_u32(msg, FRA_SUPPRESS_PREFIXLEN, rule->sup_prefixlen);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
 	if (rule->flags & IPRULE_UIDRANGE) {
 		struct fib_rule_uid_range uidrange = {
 			.start = rule->uidrange_start,
@@ -3849,6 +3852,7 @@ static int system_iprule(struct iprule *
 
 		nla_put(msg, FRA_UID_RANGE, sizeof(uidrange), &uidrange);
 	}
+#endif
 
 	if (rule->flags & IPRULE_GOTO)
 		nla_put_u32(msg, FRA_GOTO, rule->gotoid);
