--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -31,13 +31,11 @@ FIND_LIBRARY(uci NAMES uci)
 FIND_LIBRARY(ubox NAMES ubox)
 FIND_LIBRARY(ubus NAMES ubus)
 FIND_LIBRARY(json NAMES json-c json)
-FIND_LIBRARY(udebug NAMES udebug)
 FIND_LIBRARY(blobmsg_json NAMES blobmsg_json)
 
-SET(LIBS ${ubox} ${ubus} ${uci} ${json} ${blobmsg_json} ${udebug})
+SET(LIBS ${ubox} ${ubus} ${uci} ${json} ${blobmsg_json})
 
 FIND_PATH(ubox_include_dir libubox/usock.h)
-FIND_PATH(udebug_include_dir udebug.h)
 INCLUDE_DIRECTORIES(${ubox_include_dir})
 
 IF (NOT DEFINED LIBNL_LIBS)
--- a/main.c
+++ b/main.c
@@ -37,32 +37,6 @@ const char *resolv_conf = DEFAULT_RESOLV
 static char **global_argv;
 
 static struct list_head process_list = LIST_HEAD_INIT(process_list);
-static struct udebug ud;
-static struct udebug_buf udb_log;
-struct udebug_buf udb_nl;
-static const struct udebug_buf_meta meta_log = {
-	.name = "netifd_log",
-	.format = UDEBUG_FORMAT_STRING,
-};
-static const struct udebug_buf_meta meta_nl = {
-	.name = "netifd_nl",
-	.format = UDEBUG_FORMAT_PACKET,
-	.sub_format = UDEBUG_DLT_NETLINK,
-};
-static struct udebug_ubus_ring rings[] = {
-	{
-		.buf = &udb_log,
-		.meta = &meta_log,
-		.default_entries = 1024,
-		.default_size = 64 * 1024,
-	},
-	{
-		.buf = &udb_nl,
-		.meta = &meta_nl,
-		.default_entries = 1024,
-		.default_size = 64 * 1024,
-	},
-};
 
 #define DEFAULT_LOG_LEVEL L_NOTICE
 
@@ -91,42 +65,12 @@ netifd_delete_process(struct netifd_proc
 	close(proc->log.fd.fd);
 }
 
-static void __attribute__((format (printf, 1, 0)))
-netifd_udebug_vprintf(const char *format, va_list ap)
-{
-	if (!udebug_buf_valid(&udb_log))
-		return;
-
-	udebug_entry_init(&udb_log);
-	udebug_entry_vprintf(&udb_log, format, ap);
-	udebug_entry_add(&udb_log);
-}
-
-void netifd_udebug_printf(const char *format, ...)
-{
-	va_list ap;
-
-	va_start(ap, format);
-	netifd_udebug_vprintf(format, ap);
-	va_end(ap);
-}
-
-void netifd_udebug_config(struct udebug_ubus *ctx, struct blob_attr *data,
-			  bool enabled)
-{
-	udebug_ubus_apply_config(&ud, rings, ARRAY_SIZE(rings), data, enabled);
-}
-
 void
 __attribute__((format(printf, 2, 0)))
 netifd_log_message(int priority, const char *format, ...)
 {
 	va_list vl;
 
-	va_start(vl, format);
-	netifd_udebug_vprintf(format, vl);
-	va_end(vl);
-
 	if (priority > log_level)
 		return;
 
@@ -416,11 +360,6 @@ int main(int argc, char **argv)
 		openlog("netifd", 0, LOG_DAEMON);
 
 	netifd_setup_signals();
-	uloop_init();
-	udebug_init(&ud);
-	udebug_auto_connect(&ud, NULL);
-	for (size_t i = 0; i < ARRAY_SIZE(rings); i++)
-		udebug_ubus_ring_init(&ud, &rings[i]);
 
 	if (netifd_ubus_init(socket) < 0) {
 		fprintf(stderr, "Failed to connect to ubus\n");
--- a/netifd.h
+++ b/netifd.h
@@ -25,7 +25,6 @@
 #include <libubox/utils.h>
 
 #include <libubus.h>
-#include <udebug.h>
 
 #ifdef linux
 #include <netinet/ether.h>
@@ -52,7 +51,6 @@
 extern const char *resolv_conf;
 extern char *hotplug_cmd_path;
 extern unsigned int debug_mask;
-extern struct udebug_buf udb_nl;
 
 enum {
 	L_CRIT,
@@ -72,7 +70,6 @@ enum {
 #ifdef DEBUG
 #define DPRINTF(format, ...) fprintf(stderr, "%s(%d): " format, __func__, __LINE__, ## __VA_ARGS__)
 #define D(level, format, ...) do { \
-		netifd_udebug_printf("[" #level "] %s(%d): " format,  __func__, __LINE__, ## __VA_ARGS__); \
 		if (debug_mask & (1 << (DEBUG_ ## level))) { \
 			DPRINTF(format, ##__VA_ARGS__); \
 			fprintf(stderr, "\n"); \
@@ -101,10 +98,6 @@ struct netifd_process {
 	char username[32];
 };
 
-void netifd_udebug_printf(const char *format, ...)
-	__attribute__((format (printf, 1, 2)));
-void netifd_udebug_config(struct udebug_ubus *ctx, struct blob_attr *data,
-			  bool enabled);
 void netifd_log_message(int priority, const char *format, ...)
 	 __attribute__((format (printf, 2, 3)));
 
--- a/system-linux.c
+++ b/system-linux.c
@@ -211,14 +211,6 @@ abort:
 	return;
 }
 
-static void
-nl_udebug_cb(void *priv, struct nl_msg *msg)
-{
-	struct nlmsghdr *nlh = nlmsg_hdr(msg);
-
-	udebug_netlink_msg(priv, nlmsg_get_proto(msg), nlh, nlh->nlmsg_len);
-}
-
 static struct nl_sock *
 create_socket(int protocol, int groups)
 {
@@ -236,9 +228,6 @@ create_socket(int protocol, int groups)
 		return NULL;
 	}
 
-	nl_socket_set_tx_debug_cb(sock, nl_udebug_cb, &udb_nl);
-	nl_socket_set_rx_debug_cb(sock, nl_udebug_cb, &udb_nl);
-
 	return sock;
 }
 
--- a/ubus.c
+++ b/ubus.c
@@ -27,7 +27,6 @@
 struct ubus_context *ubus_ctx = NULL;
 static struct blob_buf b;
 static const char *ubus_path;
-static struct udebug_ubus udebug;
 
 /* global object */
 
@@ -1806,15 +1805,12 @@ netifd_ubus_init(const char *path)
 	netifd_add_object(&wireless_object);
 	netifd_add_iface_object();
 
-	udebug_ubus_init(&udebug, ubus_ctx, "netifd", netifd_udebug_config);
-
 	return 0;
 }
 
 void
 netifd_ubus_done(void)
 {
-	udebug_ubus_free(&udebug);
 	ubus_free(ubus_ctx);
 }
 
