From dbf5a69a1e4214753395aa79681dbb55bf7dc400 Mon Sep 17 00:00:00 2001
From: Joris Vaisvila <joris.vaisvila@teltonika.lt>
Date: Thu, 28 Nov 2024 10:47:02 +0200
Subject: [PATCH 1/1] minify .js with npx

---
 Makefile.am | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/Makefile.am b/Makefile.am
index 5c52a8b..e98b7ea 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -341,6 +341,9 @@ clean-demo:
 	     -e "s/^#/\/\/ #/"                                                \
 	     -e "s/VERSION/\"@VERSION@ @VCS_REVISION@\"/g"                    \
 	     "$<" >"$@"
+	@mv "$@" "$@.big.js"
+	@npx esbuild "$@.big.js" --minify --outfile="$@"
+	@rm "$@.big.js"
 
 .js.h:
 	@echo od "$<" '>'"$@"
-- 
2.47.0

