Index: openvpn-2.6.14/src/openvpn/misc.h
===================================================================
--- openvpn-2.6.14.orig/src/openvpn/misc.h
+++ openvpn-2.6.14/src/openvpn/misc.h
@@ -66,7 +66,7 @@ struct user_pass
 #ifdef ENABLE_PKCS11
 #define USER_PASS_LEN 4096
 #else
-#define USER_PASS_LEN 128
+#define USER_PASS_LEN 512
 #endif
     /* Note that username and password are expected to be null-terminated */
     char username[USER_PASS_LEN];
