Index: openvpn-2.6.9/src/openvpn/forward.c
===================================================================
--- openvpn-2.6.9.orig/src/openvpn/forward.c
+++ openvpn-2.6.9/src/openvpn/forward.c
@@ -1241,6 +1241,8 @@ process_incoming_dco(struct context *c)
     switch (dco->dco_message_type)
     {
         case OVPN_CMD_DEL_PEER:
+            /* peer is gone, unset ID to prevent more kernel calls */
+            c->c2.tls_multi->dco_peer_id = -1;
             if (dco->dco_del_peer_reason == OVPN_DEL_PEER_REASON_EXPIRED)
             {
                 msg(D_DCO_DEBUG, "%s: received peer expired notification of for peer-id "
