--- a/lib/packet_mosq.c
+++ b/lib/packet_mosq.c
@@ -484,6 +484,11 @@
 				}
 				break;
 		}
 
+		if (mosq->in_packet.remaining_length >= db.config->max_packet_size &&
+		    db.config->max_packet_size != 0) {
+			packet__cleanup(&mosq->in_packet);
+                        return MOSQ_ERR_OVERSIZE_PACKET;
+		}
 		if(db.config->max_packet_size > 0 && mosq->in_packet.remaining_length+1 > db.config->max_packet_size){
 			if(mosq->protocol == mosq_p_mqtt5){
 				send__disconnect(mosq, MQTT_RC_PACKET_TOO_LARGE, NULL);
