Index: linux-5.10.191/drivers/net/ethernet/mediatek/mtk_eth_soc.c
===================================================================
--- linux-5.10.191.orig/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ linux-5.10.191/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -1088,7 +1088,12 @@ static int mtk_tx_map(struct sk_buff *sk
 			txd_pdma->txd2 |= TX_DMA_LS1;
 	}
 
-	netdev_sent_queue(dev, skb->len);
+	for (i = 0; i < MTK_MAC_COUNT; i++) {
+		if (!eth->netdev[i])
+			continue;
+		netdev_sent_queue(eth->netdev[i], tx_num);
+	}
+
 	skb_tx_timestamp(skb);
 
 	ring->next_free = mtk_qdma_phys_to_virt(ring, txd->txd2);
@@ -1442,7 +1447,7 @@ static int mtk_poll_tx_qdma(struct mtk_e
 			break;
 
 		if (skb != (struct sk_buff *)MTK_DMA_DUMMY_DESC) {
-			bytes[mac] += skb->len;
+			bytes[0] += mtk_cal_txd_req(skb);
 			done[mac]++;
 			budget--;
 		}
@@ -1515,9 +1520,9 @@ static int mtk_poll_tx(struct mtk_eth *e
 		budget = mtk_poll_tx_pdma(eth, budget, done, bytes);
 
 	for (i = 0; i < MTK_MAC_COUNT; i++) {
-		if (!eth->netdev[i] || !done[i])
+		if (!eth->netdev[i])
 			continue;
-		netdev_completed_queue(eth->netdev[i], done[i], bytes[i]);
+		netdev_completed_queue(eth->netdev[i], done[i], bytes[0]);
 		total += done[i];
 		eth->tx_packets += done[i];
 		eth->tx_bytes += bytes[i];
