Index: linux-5.10.191/drivers/net/ethernet/mediatek/mtk_eth_soc.h
===================================================================
--- linux-5.10.191.orig/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ linux-5.10.191/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -23,7 +23,7 @@
 #define MTK_MAX_RX_LENGTH	1536
 #define MTK_MAX_RX_LENGTH_2K	2048
 #define MTK_TX_DMA_BUF_LEN	0x3fff
-#define MTK_DMA_SIZE		512
+#define MTK_DMA_SIZE		2048
 #define MTK_NAPI_WEIGHT		64
 #define MTK_MAC_COUNT		2
 #define MTK_RX_ETH_HLEN		(ETH_HLEN + ETH_FCS_LEN)
