Index: coova-chilli-2025-01-09/src/linux/xt_coova.c
===================================================================
--- coova-chilli-2025-01-09.orig/src/linux/xt_coova.c	2025-01-09 11:43:53.000000000 +0200
+++ coova-chilli-2025-01-09/src/linux/xt_coova.c	2025-01-21 10:27:23.509316414 +0200
@@ -479,10 +479,10 @@
 	if (st == NULL)
 		return -ENOMEM;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
-	st->table = PDE_DATA(inode);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,17,0)
+	st->table = pde_data(inode);
 #else
-	st->table = pde->data;
+	st->table = PDE_DATA(inode);
 #endif
 	return 0;
 }
@@ -491,11 +491,10 @@
 coova_mt_proc_write(struct file *file, const char __user *input,
 		    size_t size, loff_t *loff)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
-	struct coova_table *t = PDE_DATA(file_inode(file));
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,17,0)
+	struct coova_table *t = pde_data(file_inode(file));
 #else
-	const struct proc_dir_entry *pde = PDE(file->f_path.dentry->d_inode);
-	struct coova_table *t = pde->data;
+	struct coova_table *t = PDE_DATA(file_inode(file));
 #endif
 	struct coova_entry *e;
 	char buf[sizeof("+b335:1d35:1e55:dead:c0de:1715:5afe:c0de")];
