Index: libdlms-b5395c98/development/include/gxignore.h
===================================================================
--- libdlms-b5395c98.orig/development/include/gxignore.h
+++ libdlms-b5395c98/development/include/gxignore.h
@@ -43,7 +43,7 @@ extern "C" {
     // #define DLMS_IGNORE_PLC
     // #define DLMS_IGNORE_IEC
     // #define DLMS_IGNORE_NOTIFY
-    // #define DLMS_IGNORE_SERVER
+    #define DLMS_IGNORE_SERVER
     // #define DLMS_IGNORE_CLIENT
     // #define GX_DLMS_MICROCONTROLLER
     // #define DLMS_IGNORE_HIGH_SHA256
@@ -157,7 +157,7 @@ extern "C" {
 
     // Use 32 bit max size bytebuffer instead of 16 bit.
     // This might be used in client side if a lot of data is read from the meter.
-    // #define GX_DLMS_BYTE_BUFFER_SIZE_32
+     #define GX_DLMS_BYTE_BUFFER_SIZE_32
 
     //Serializer is not used.
     //DLMS_IGNORE_SERIALIZER
