Index: libdlms-fa855d2e/development/src/gxsetmalloc.c
===================================================================
--- libdlms-fa855d2e.orig/development/src/gxsetmalloc.c
+++ libdlms-fa855d2e/development/src/gxsetmalloc.c
@@ -522,6 +522,12 @@ int updateDayProfileTable(dlmsSettings*
             {
                 break;
             }
+
+		if (!tmp2->byteArr) {
+			gxfree(ac);
+			break;
+		}
+
             if (tmp2->vt == DLMS_DATA_TYPE_OCTET_STRING && tmp2->byteArr != NULL)
             {
 #ifndef DLMS_IGNORE_OBJECT_POINTERS
