diff --git a/development/src/dlms.c b/development/src/dlms.c
index 2894e57..d539d08 100644
--- a/development/src/dlms.c
+++ b/development/src/dlms.c
@@ -3750,17 +3750,10 @@ unsigned char dlms_isPlcSfskData(gxByteBuffer* buff)
 }
 #endif //DLMS_IGNORE_PLC
 
-#if defined(GX_DLMS_BYTE_BUFFER_SIZE_32) || (!defined(GX_DLMS_MICROCONTROLLER) && (defined(_WIN32) || defined(_WIN64) || defined(__linux__)))
 int dlms_getDataFromBlock(gxByteBuffer* data, uint32_t index)
-#else
-int dlms_getDataFromBlock(gxByteBuffer* data, uint16_t index)
-#endif
 {
-#if defined(GX_DLMS_BYTE_BUFFER_SIZE_32) || (!defined(GX_DLMS_MICROCONTROLLER) && (defined(_WIN32) || defined(_WIN64) || defined(__linux__)))
     uint32_t pos, len = data->position - index;
-#else
-    uint16_t pos, len = data->position - index;
-#endif
+
     if (data->size == data->position)
     {
         bb_clear(data);
@@ -6385,7 +6378,7 @@ void dlms_applyHdlcData(
         settings->clientAddress = result->source;
     }
 
-    if (!reply->preEstablished) {
+    if (!reply->preEstablished && *frame != 0x13) {
         // If U frame.
         if ((*frame & 0x3) == 3)
         {
@@ -6409,7 +6402,6 @@ void dlms_applyHdlcData(
         {
             settings->receiverFrame = *frame;
         }
-
     }
 }
 
@@ -6442,6 +6434,20 @@ int dlms_getData3(
             result.moreData = reply->moreData;
             result.packetLength = reply->packetLength;
             ret = dlms_getHdlcData(settings, recv_data, &result, reply->preEstablished, first);
+
+            // Ignore data unnumbered information packets, because for SOME REASON
+            // they mess up the "Send sequence number" which makes it "desync".
+            //
+            // I couldn't track down the real cause of this "desync", so it is just simpler to ignore
+            // these packets, eithe way DLMS Client isin't using these pacekts in any way, so for you use
+            // case this will be fine.
+            //
+            // Keep the DLMS Client running for long enough until the example tcp server sends data
+            // notification message.
+            if (result.frameType == 0x13) {
+                continue;
+            }
+
             if (ret == DLMS_ERROR_CODE_OK && result.complete) {
                 dlms_applyHdlcData(&result, settings, reply, &frame);
             }
