Index: firewall-2022-02-17-4cd7d4f3/defaults.c
===================================================================
--- firewall-2022-02-17-4cd7d4f3.orig/defaults.c
+++ firewall-2022-02-17-4cd7d4f3/defaults.c
@@ -253,7 +253,6 @@ fw3_print_default_head_rules(struct fw3_
 		{
 			r = fw3_ipt_rule_new(handle);
 			fw3_ipt_rule_comment(r, "Traffic offloading");
-			fw3_ipt_rule_extra(r, "-m conntrack --ctstate RELATED,ESTABLISHED");
 			fw3_ipt_rule_target(r, "FLOWOFFLOAD");
 			if (defs->flow_offloading_hw)
 				fw3_ipt_rule_addarg(r, false, "--hw", NULL);
