--- a/iwinfo.c
+++ b/iwinfo.c
@@ -452,14 +452,22 @@ rpc_iwinfo_scan(struct ubus_context *ctx
 			if (e->ht_chan_info.primary_chan) {
 				t = blobmsg_open_table(&buf, "ht_operation");
 				blobmsg_add_u32(&buf, "primary_channel", e->ht_chan_info.primary_chan);
-				blobmsg_add_string(&buf, "secondary_channel_offset", ht_secondary_offset[e->ht_chan_info.secondary_chan_off]);
-				blobmsg_add_u32(&buf, "channel_width", ht_chan_width[e->ht_chan_info.chan_width]);
+			
+				blobmsg_add_string(&buf, "secondary_channel_offset", e->ht_chan_info.secondary_chan_off > ARRAY_SIZE(ht_secondary_offset) ?
+				"unknown" : ht_secondary_offset[e->ht_chan_info.secondary_chan_off]);
+
+				blobmsg_add_u32(&buf, "channel_width", e->ht_chan_info.chan_width > ARRAY_SIZE(ht_chan_width) ?
+				0 : ht_chan_width[e->ht_chan_info.chan_width]);
+
 				blobmsg_close_table(&buf, t);
 			}
 
 			if (e->vht_chan_info.center_chan_1) {
 				t = blobmsg_open_table(&buf, "vht_operation");
-				blobmsg_add_u32(&buf, "channel_width", vht_chan_width[e->vht_chan_info.chan_width]);
+
+				blobmsg_add_u32(&buf, "channel_width", e->vht_chan_info.chan_width > ARRAY_SIZE(vht_chan_width) ?
+				0 : vht_chan_width[e->vht_chan_info.chan_width]);
+
 				blobmsg_add_u32(&buf, "center_freq_1", e->vht_chan_info.center_chan_1);
 				blobmsg_add_u32(&buf, "center_freq_2", e->vht_chan_info.center_chan_2);
 				blobmsg_close_table(&buf, t);
