Index: rpcd-2021-03-11-ccb75178/sys.c
===================================================================
--- rpcd-2021-03-11-ccb75178.orig/sys.c
+++ rpcd-2021-03-11-ccb75178/sys.c
@@ -340,6 +340,23 @@ rpc_sys_reboot(struct ubus_context *ctx,
 	return 0;
 }
 
+static int
+rpc_sys_poweroff(struct ubus_context *ctx, struct ubus_object *obj,
+                 struct ubus_request_data *req, const char *method,
+                 struct blob_attr *msg)
+{
+	if (!fork()) {
+		sleep(2);
+		int result = system("/sbin/poweroff");
+		if (result == -1) {
+			return rpc_errno_status();
+		}
+		exit(result);
+	}
+
+	return 0;
+}
+
 static struct ubus_object
 *rpc_sys_api_init(const struct rpc_daemon_ops *o, struct ubus_context *ctx)
 {
@@ -352,6 +369,7 @@ static struct ubus_object
 		UBUS_METHOD_NOARG("upgrade_clean", rpc_sys_upgrade_clean),
 		UBUS_METHOD_NOARG("factory", rpc_sys_factory),
 		UBUS_METHOD_NOARG("reboot", rpc_sys_reboot),
+		UBUS_METHOD_NOARG("poweroff", rpc_sys_poweroff),
 	};
 
 	static struct ubus_object_type sys_type =
