diff --git a/plugin.c b/plugin.c
index d0e8a2f..9c803b3 100644
--- a/plugin.c
+++ b/plugin.c
@@ -763,8 +763,10 @@ static int rpc_plugin_register_library(struct ubus_context *ctx, const char *pat
 	void *dlh;
 
 	dlh = dlopen(path, RTLD_LAZY | RTLD_LOCAL);
-	if (!dlh)
+	if (!dlh) {
+		LOG_ERROR("Failed to open plugin %s: %s", path, dlerror());
 		return UBUS_STATUS_UNKNOWN_ERROR;
+	}
 
 	p = dlsym(dlh, "rpc_plugin");
 	if (!p) {
