Index: rpcd-2021-03-11-ccb75178/main.c
===================================================================
--- rpcd-2021-03-11-ccb75178.orig/main.c
+++ rpcd-2021-03-11-ccb75178/main.c
@@ -30,10 +30,14 @@
 #include <rpcd/rc.h>
 #include <rpcd/session.h>
 #include <rpcd/uci.h>
+
 #define RPC_UCI_DIR_PREFIX	"/var/run/rpcd"
 #define RPC_UCI_SAVEDIR_PREFIX	RPC_UCI_DIR_PREFIX "/uci-"
+#define RPCD_CONNECT_RETRY	10
+#define RPCD_RECONNECT_DELAY 100000
 static struct ubus_auto_conn conn;
 static bool respawn = false;
+static bool ubus_connected = false;
 
 int rpc_exec_timeout = RPC_EXEC_DEFAULT_TIMEOUT;
 
@@ -112,11 +116,28 @@ exec_self(int argc, char **argv)
 
 static void ubus_connect_handler(struct ubus_context *ctx)
 {
+	ubus_connected = true;
 	rpc_session_api_init(ctx);
 	rpc_rc_api_init(ctx);
 	rpc_plugin_api_init(ctx);
 }
 
+static int ubus_connect_retry(struct ubus_auto_conn *conn, const char * ubus_socket) {
+	int retries = 0;
+
+	do {
+		ubus_connected = false;
+		conn->cb = ubus_connect_handler;
+		conn->path = ubus_socket;
+		ubus_auto_connect(conn);
+		if (ubus_connected)
+			return 0;
+		usleep(RPCD_RECONNECT_DELAY);
+		retries++;
+	} while (retries < RPCD_CONNECT_RETRY);
+	return -1;
+}
+
 int main(int argc, char **argv)
 {
 	struct stat s;
@@ -155,9 +176,6 @@ int main(int argc, char **argv)
 
 	uloop_init();
 
-	conn.cb	  = ubus_connect_handler;
-	conn.path = ubus_socket;
-
 	rpc_session_init();
 
 	hangup = getenv("RPC_HANGUP");
@@ -167,7 +185,11 @@ int main(int argc, char **argv)
 	else
 		rpc_session_thaw();
 
-	ubus_auto_connect(&conn);
+	if (ubus_connect_retry(&conn, ubus_socket) != 0) {
+		fprintf(stderr, "Failed to connect to ubus\n");
+		return -1;
+	}
+
 	uloop_run();
 	uloop_done();
 	ubus_auto_shutdown(&conn);
