Index: rpcd-2021-03-11-ccb75178/session.c
===================================================================
--- rpcd-2021-03-11-ccb75178.orig/session.c
+++ rpcd-2021-03-11-ccb75178/session.c
@@ -36,6 +36,10 @@
 #include <shadow.h>
 #endif
 
+#ifndef LOGIN_NAME_MAX
+#define LOGIN_NAME_MAX sysconf(_SC_LOGIN_NAME_MAX)
+#endif
+
 #include <rpcd/session.h>
 
 #ifdef ENABLE_PAM_SUPPORT
@@ -1167,6 +1171,9 @@ rpc_login_test_login(struct uci_context
 	struct uci_section *s = NULL;
 	struct uci_element *e = NULL;
 	struct uci_ptr ptr    = { .package = "rpcd" };
+	// Max length of $p$ + LOGIN_NAME_MAX + null terminator
+	char user_shadow_entry[LOGIN_NAME_MAX + 4] = { 0 };
+	snprintf(user_shadow_entry, sizeof(user_shadow_entry), "$p$%s", username);
 
 	if (!uci_lookup_ptr(uci, &ptr, NULL, false) && ptr.p) {
 		uci_unload(uci, ptr.p);
@@ -1236,19 +1243,6 @@ rpc_login_test_login(struct uci_context
 			break;
 		}
 
-		/* test for matching password */
-		ptr.option = "password";
-		ptr.o = NULL;
-
-		if (uci_lookup_ptr(uci, &ptr, NULL, true))
-			continue;
-
-		if (!ptr.o)
-			continue;
-
-		if (ptr.o->type != UCI_TYPE_STRING)
-			continue;
-
 #ifdef ENABLE_PAM_SUPPORT
 		int auth_type = rpc_login_test_auth_type(uci, s);
 		if (auth_type & AUTH_TYPE_PAM) {
@@ -1263,7 +1257,7 @@ rpc_login_test_login(struct uci_context
 		} else if (auth_type == AUTH_TYPE_NONE || auth_type & AUTH_TYPE_SHADOW)
 #endif
 		{
-			if (rpc_login_test_password(ptr.o->v.string, password)) {
+			if (rpc_login_test_password(user_shadow_entry, password)) {
 				ptr.option = "group";
 				ptr.o = NULL;
 
