--- a/block.c
+++ b/block.c
@@ -1570,10 +1570,22 @@ static int mount_extroot(char *cfg)
 		return -1;
 	}
 
+#ifdef RUT9_PLATFORM
+	char kmod_loader[64];
+	snprintf(kmod_loader, sizeof(kmod_loader),
+	         "/sbin/kmodloader /etc/modules-late.d/mmc-spi");
+	if (system(kmod_loader))
+		ULOG_ERR("extroot: couldn't kmodload mmc\n");
+#endif
+
 	/* Find block device pointed by the mount config */
 	pr = find_block_info(m->uuid, m->label, m->device);
 
+#ifdef RUT9_PLATFORM
+	for (unsigned c = 10, delay_root = 1; !pr && c; c--) {
+#else
 	if (!pr && delay_root){
+#endif
 		ULOG_INFO("extroot: device not present, retrying in %u seconds\n", delay_root);
 		sleep(delay_root);
 		make_devs();
