--- a/libfstools/overlay.c
+++ b/libfstools/overlay.c
@@ -37,6 +37,7 @@
 #define SWITCH_JFFS2 "/tmp/.switch_jffs2"
 #define OVERLAYDIR "/rom/overlay"
 
+static void handle_glob(struct stat *s, glob_t *gl, int (*cb)(const char*));
 static bool keep_sysupgrade;
 
 static int
@@ -70,6 +71,39 @@ handle_rmdir(const char *dir)
 	return 0;
 }
 
+static void
+handle_glob(struct stat *s, glob_t *gl, int (*cb)(const char*))
+{
+	for (int i = 0; i < gl->gl_pathc; i++) {
+		char *dir = gl->gl_pathv[i];
+		int len = strlen(gl->gl_pathv[i]);
+		int err;
+
+		/* Quick way of skipping files */
+		if (dir[len - 1] != '/')
+			continue;
+
+		/* Glob with /.* searches for current and parent directories, skip them */
+		if (strstr(dir, "/./") || strstr(dir, "/../")) {
+                	continue;
+            	}
+
+		// /* lstat needs path without a trailing slash */
+		if (len > 1) {
+			dir[len - 1] = '\0';
+		}
+
+		err = lstat(gl->gl_pathv[i], s);
+		if (len > 1) {
+			dir[len - 1] = '/';
+		}
+
+		if (!err && !S_ISLNK(s->st_mode)) {
+			foreachdir(gl->gl_pathv[i], cb);
+		}
+	}
+}
+
 void
 foreachdir(const char *dir, int (*cb)(const char*))
 {
@@ -78,7 +112,6 @@ foreachdir(const char *dir, int (*cb)(co
 	struct stat s = { 0 };
 	size_t dirlen = strlen(dir);
 	glob_t gl;
-	int j;
 
 	if (dirlen + sizeof("/*") > globdirlen) {
 		/* Alloc extra 256 B to avoid too many reallocs */
@@ -93,28 +126,16 @@ foreachdir(const char *dir, int (*cb)(co
 	}
 
 	sprintf(globdir, "%s/*", dir);
-
 	/* Include GLOB_MARK as callbacks expect a trailing slash */
-	if (!glob(globdir, GLOB_NOESCAPE | GLOB_MARK | GLOB_ONLYDIR, NULL, &gl))
-		for (j = 0; j < gl.gl_pathc; j++) {
-			char *dir = gl.gl_pathv[j];
-			int len = strlen(gl.gl_pathv[j]);
-			int err;
-
-			/* Quick way of skipping files */
-			if (dir[len - 1] != '/')
-				continue;
-
-			/* lstat needs path without a trailing slash */
-			if (len > 1)
-				dir[len - 1] = '\0';
-			err = lstat(gl.gl_pathv[j], &s);
-			if (len > 1)
-				dir[len - 1] = '/';
+	if (!glob(globdir, GLOB_NOESCAPE | GLOB_MARK | GLOB_ONLYDIR, NULL, &gl)) {
+		handle_glob(&s, &gl, cb);
+	}
 
-			if (!err && !S_ISLNK(s.st_mode))
-				foreachdir(gl.gl_pathv[j], cb);
+	sprintf(globdir, "%s/.*", dir);
+	if (!glob(globdir, GLOB_NOESCAPE | GLOB_MARK | GLOB_ONLYDIR, NULL, &gl)) {
+		handle_glob(&s, &gl, cb);
 	}
+
 	cb(dir);
 }
 
