--- a/libfstools/common.c
+++ b/libfstools/common.c
@@ -91,24 +91,6 @@ int block_file_identify(FILE *f, uint64_
 	return FS_NONE;
 }
 
-static bool use_f2fs(struct volume *v, uint64_t offset, const char *bdev)
-{
-	uint64_t size = 0;
-	bool ret = false;
-	int fd;
-
-	fd = open(bdev, O_RDONLY);
-	if (fd < 0)
-		return false;
-
-	if (ioctl(fd, BLKGETSIZE64, &size) == 0)
-		ret = size - offset > F2FS_MINSIZE;
-
-	close(fd);
-
-	return ret;
-}
-
 int block_volume_format(struct volume *v, uint64_t offset, const char *bdev)
 {
 	int ret = 0;
@@ -160,10 +142,7 @@ int block_volume_format(struct volume *v
 		}
 do_format:
 		ULOG_INFO("overlay filesystem in %s has not been formatted yet\n", v->blk);
-		if (use_f2fs(v, offset, bdev))
-			snprintf(str, sizeof(str), "mkfs.f2fs -q -f -l rootfs_data %s", v->blk);
-		else
-			snprintf(str, sizeof(str), "mkfs.ext4 -q -F -L rootfs_data %s", v->blk);
+		snprintf(str, sizeof(str), "mkfs.ext4 -q -F -L rootfs_data %s", v->blk);
 
 		ret = system(str);
 		break;
