Index: fstools-2022-05-03-9e11b372/block.c
===================================================================
--- fstools-2022-05-03-9e11b372.orig/block.c
+++ fstools-2022-05-03-9e11b372/block.c
@@ -1492,9 +1492,13 @@ static int check_extroot(char *path)
 		libubi_t libubi;
 
 		libubi = libubi_open();
+		if (!libubi) {
+			goto skip_ubi;
+		}
 		err = find_block_ubi_RO(libubi, "rootfs", devpath, sizeof(devpath));
 		libubi_close(libubi);
 	}
+skip_ubi:
 #endif
 	if (err) {
 		err = find_root_dev(devpath, sizeof(devpath));
@@ -1742,6 +1747,9 @@ static int main_extroot(int argc, char *
 	/* ... but it also could be an UBI volume */
 	memset(blkdev_path, 0, sizeof(blkdev_path));
 	libubi = libubi_open();
+	if (!libubi) {
+		goto skip_ubi;
+	}
 	find_block_ubi(libubi, "rootfs_data", blkdev_path, sizeof(blkdev_path));
 	libubi_close(libubi);
 	if (blkdev_path[0]) {
@@ -1758,6 +1767,7 @@ static int main_extroot(int argc, char *
 		rmdir(cfg);
 		return err;
        }
+skip_ubi:
 #endif
 
 	/* As a last resort look for /etc/config/fstab on "rootfs" partition */
