--- a/drivers/net/wireless/ath/ath10k/wmi.c
+++ b/drivers/net/wireless/ath/ath10k/wmi.c
@@ -6358,6 +6358,10 @@ static void ath10k_wmi_10_4_op_rx(struct
 	case WMI_10_4_PEER_RATECODE_LIST_EVENTID:
 	case WMI_10_4_WDS_PEER_EVENTID:
 	case WMI_10_4_DEBUG_FATAL_CONDITION_EVENTID:
+	case WMI_10_4_PDEV_CTL_FAILSAFE_CHECK_EVENTID:
+	case WMI_10_4_PEER_TX_PN_RESPONSE_EVENTID:
+	case WMI_10_4_VDEV_DELETE_ALL_PEER_RESP_EVENTID:
+	case WMI_10_4_PDEV_MULTIPLE_VDEV_RESTART_RESP_EVENTID:
 		ath10k_dbg(ar, ATH10K_DBG_WMI,
 			   "received event id %d not implemented\n", id);
 		break;
--- a/drivers/net/wireless/ath/ath10k/wmi.h
+++ b/drivers/net/wireless/ath/ath10k/wmi.h
@@ -1967,6 +1967,10 @@ enum wmi_10_4_event_id {
 	WMI_10_4_HOST_SWFDA_EVENTID,
 	WMI_10_4_ESP_ESTIMATE_EVENTID,
 	WMI_10_4_DFS_STATUS_CHECK_EVENTID,
+	WMI_10_4_PDEV_CTL_FAILSAFE_CHECK_EVENTID,
+	WMI_10_4_PEER_TX_PN_RESPONSE_EVENTID,
+	WMI_10_4_VDEV_DELETE_ALL_PEER_RESP_EVENTID,
+	WMI_10_4_PDEV_MULTIPLE_VDEV_RESTART_RESP_EVENTID,
 	WMI_10_4_PDEV_UTF_EVENTID = WMI_10_4_END_EVENTID - 1,
 };
 
