--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -2981,6 +2981,7 @@ int wpa_wps_supplicant_fast_associate(st
 static void wnm_bss_keep_alive(void *eloop_ctx, void *sock_ctx)
 {
 	struct wpa_supplicant *wpa_s = eloop_ctx;
+	u8 buf = 0;
 
 	if (wpa_s->wpa_state < WPA_ASSOCIATED)
 		return;
@@ -2993,7 +2994,7 @@ static void wnm_bss_keep_alive(void *elo
 		 * this */
 		if (wpa_s->l2)
 			l2_packet_send(wpa_s->l2, wpa_s->bssid, 0x0800,
-				       (u8 *) "", 0);
+				       &buf, sizeof(buf));
 	}
 
 #ifdef CONFIG_SME
