--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -472,21 +472,21 @@ static int acs_survey_is_sufficient(stru
 {
 	if (!(survey->filled & SURVEY_HAS_NF)) {
 		survey->nf = -95;
-		wpa_printf(MSG_INFO,
+		wpa_printf(MSG_DEBUG,
 			   "ACS: Survey for freq %d is missing noise floor",
 			   survey->freq);
 	}
 
 	if (!(survey->filled & SURVEY_HAS_CHAN_TIME)) {
 		survey->channel_time = 0;
-		wpa_printf(MSG_INFO,
+		wpa_printf(MSG_DEBUG,
 			   "ACS: Survey for freq %d is missing channel time",
 			   survey->freq);
 	}
 
 	if (!(survey->filled & SURVEY_HAS_CHAN_TIME_BUSY) &&
 	    !(survey->filled & SURVEY_HAS_CHAN_TIME_RX)) {
-		wpa_printf(MSG_INFO,
+		wpa_printf(MSG_DEBUG,
 			   "ACS: Survey for freq %d is missing RX and busy time (at least one is required)",
 			   survey->freq);
 	}
