--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -2548,9 +2548,15 @@ static int ieee802_11_allowed_address(st
 	res = hostapd_allowed_address(hapd, addr, msg, len, info, 0);
 
 	if (res == HOSTAPD_ACL_REJECT) {
-		wpa_printf(MSG_DEBUG, "Station " MACSTR
-			   " not allowed to authenticate",
-			   MAC2STR(addr));
+		if (!hapd->conf || !hapd->iconf || !hapd->iconf->bss[0]) {
+			return HOSTAPD_ACL_REJECT;
+		}
+
+		// Check syslog level from phy*.conf file
+		if (hapd->iconf->bss[0]->logger_syslog_level >= 3) {
+			wpa_printf(MSG_INFO, "Station " MACSTR " not allowed to authenticate to '%s'", MAC2STR(addr),
+				   wpa_ssid_txt(hapd->conf->ssid.ssid, hapd->conf->ssid.ssid_len));
+		}
 		return HOSTAPD_ACL_REJECT;
 	}
 
