Index: hostapd-2020-06-08-5a8b3662/wpa_supplicant/sme.c
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/wpa_supplicant/sme.c
+++ hostapd-2020-06-08-5a8b3662/wpa_supplicant/sme.c
@@ -918,6 +918,13 @@ no_fils:
 		wpa_supplicant_mark_disassoc(wpa_s);
 		wpabuf_free(resp);
 		wpas_connect_work_done(wpa_s);
+		// The driver can fail with incorrect channel configuration
+		// without the network re-add the driver will be stuck on bad cfg
+		if (!wpa_s->current_ssid)
+			return;
+
+		wpa_supplicant_remove_network(wpa_s, wpa_s->current_ssid->id);
+		wpa_supplicant_add_network(wpa_s);
 		return;
 	}
 
