--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -167,7 +167,7 @@
 }
 
 static void
-hostapd_bss_ban_client(struct hostapd_data *hapd, u8 *addr, int time)
+hostapd_bss_ban_client(struct hostapd_data *hapd, u8 *addr, int time, bool lifetime)
 {
 	struct ubus_banned_client *ban;
 
@@ -176,7 +176,7 @@
 
 	ban = avl_find_element(&hapd->ubus.banned, addr, ban, avl);
 	if (!ban) {
-		if (!time)
+		if (!time && !lifetime)
 			return;
 
 		ban = os_zalloc(sizeof(*ban));
@@ -184,14 +184,17 @@
 		ban->avl.key = ban->addr;
 		avl_insert(&hapd->ubus.banned, &ban->avl);
 	} else {
-		eloop_cancel_timeout(hostapd_bss_del_ban, ban, hapd);
-		if (!time) {
-			hostapd_bss_del_ban(ban, hapd);
-			return;
+		if(!lifetime) {
+			eloop_cancel_timeout(hostapd_bss_del_ban, ban, hapd);
+			if (!time) {
+				hostapd_bss_del_ban(ban, hapd);
+				return;
+			}
 		}
 	}
 
-	eloop_register_timeout(0, time * 1000, hostapd_bss_del_ban, ban, hapd);
+	if(!lifetime)
+		eloop_register_timeout(0, time * 1000, hostapd_bss_del_ban, ban, hapd);
 }
 
 static int
@@ -447,6 +450,7 @@
 	DEL_CLIENT_REASON,
 	DEL_CLIENT_DEAUTH,
 	DEL_CLIENT_BAN_TIME,
+	DEL_CLIENT_LIFETIME,
 	__DEL_CLIENT_MAX
 };
 
@@ -455,6 +459,7 @@
 	[DEL_CLIENT_REASON] = { "reason", BLOBMSG_TYPE_INT32 },
 	[DEL_CLIENT_DEAUTH] = { "deauth", BLOBMSG_TYPE_INT8 },
 	[DEL_CLIENT_BAN_TIME] = { "ban_time", BLOBMSG_TYPE_INT32 },
+	[DEL_CLIENT_LIFETIME] = { "lifetime", BLOBMSG_TYPE_INT8 },
 };
 
 static int
@@ -466,6 +471,8 @@
 	struct hostapd_data *hapd = container_of(obj, struct hostapd_data, ubus.obj);
 	struct sta_info *sta;
 	bool deauth = false;
+	int ban_time = 0;
+	bool lifetime = false;
 	int reason;
 	u8 addr[ETH_ALEN];
 
@@ -482,6 +489,12 @@
 
 	if (tb[DEL_CLIENT_DEAUTH])
 		deauth = blobmsg_get_bool(tb[DEL_CLIENT_DEAUTH]);
+
+	if (tb[DEL_CLIENT_BAN_TIME])
+		ban_time = blobmsg_get_u32(tb[DEL_CLIENT_BAN_TIME]);
+
+	if (tb[DEL_CLIENT_LIFETIME])
+		lifetime = blobmsg_get_bool(tb[DEL_CLIENT_LIFETIME]);
 
 	sta = ap_get_sta(hapd, addr);
 	if (sta) {
@@ -494,8 +507,7 @@
 		}
 	}
 
-	if (tb[DEL_CLIENT_BAN_TIME])
-		hostapd_bss_ban_client(hapd, addr, blobmsg_get_u32(tb[DEL_CLIENT_BAN_TIME]));
+	hostapd_bss_ban_client(hapd, addr, ban_time, lifetime);
 
 	return 0;
 }

