Index: hostapd-2020-06-08-5a8b3662/src/ap/ubus.c
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/src/ap/ubus.c
+++ hostapd-2020-06-08-5a8b3662/src/ap/ubus.c
@@ -679,6 +679,11 @@ static const struct blobmsg_policy confi
 	[CONFIG_FILE] = { "config", BLOBMSG_TYPE_STRING },
 };
 
+static const char *add_error_t[] = {
+	[HOSTAPD_INIT_UNKNOWN] = "unknown error",
+	[HOSTAPD_INIT_IN_USE]  = "already in use",
+};
+
 static int
 hostapd_config_add(struct ubus_context *ctx, struct ubus_object *obj,
 		   struct ubus_request_data *req, const char *method,
@@ -687,6 +692,7 @@ hostapd_config_add(struct ubus_context *
 	struct blob_attr *tb[__CONFIG_MAX];
 	struct hapd_interfaces *interfaces = get_hapd_interfaces_from_object(obj);
 	char buf[128];
+	int ret = 0;
 
 	blobmsg_parse(config_add_policy, __CONFIG_MAX, tb, blob_data(msg), blob_len(msg));
 
@@ -697,8 +703,13 @@ hostapd_config_add(struct ubus_context *
 		blobmsg_get_string(tb[CONFIG_IFACE]),
 		blobmsg_get_string(tb[CONFIG_FILE]));
 
-	if (hostapd_add_iface(interfaces, buf))
+	if ((ret = hostapd_add_iface(interfaces, buf))) {
+		blob_buf_init(&b, 0);
+		blobmsg_add_string(&b, "error", add_error_t[ret]);
+		blobmsg_add_u32(&b, "errno", ret);
+		ubus_send_reply(ctx, req, b.head);
 		return UBUS_STATUS_INVALID_ARGUMENT;
+	}
 
 	blob_buf_init(&b, 0);
 	blobmsg_add_u32(&b, "pid", getpid());
Index: hostapd-2020-06-08-5a8b3662/src/ap/hostapd.h
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/src/ap/hostapd.h
+++ hostapd-2020-06-08-5a8b3662/src/ap/hostapd.h
@@ -90,6 +90,13 @@ enum hostapd_chan_status {
 	HOSTAPD_CHAN_ACS = 2, /* ACS work being performed */
 };
 
+enum hostapd_init_rsp {
+	HOSTAPD_INIT_OK,
+	HOSTAPD_INIT_UNKNOWN,
+	HOSTAPD_INIT_IN_USE,
+	__HOSTAPD_INIT_MAX,
+};
+
 struct hostapd_probereq_cb {
 	int (*cb)(void *ctx, const u8 *sa, const u8 *da, const u8 *bssid,
 		  const u8 *ie, size_t ie_len, int ssi_signal);
@@ -618,7 +625,7 @@ struct hostapd_iface * hostapd_init(stru
 				    const char *config_file);
 struct hostapd_iface *
 hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
-			   const char *config_fname, int debug);
+			   const char *config_fname, int debug, enum hostapd_init_rsp *init_rsp);
 void hostapd_new_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta,
 			   int reassoc);
 void hostapd_interface_deinit_free(struct hostapd_iface *iface);
@@ -628,7 +635,7 @@ int hostapd_disable_iface(struct hostapd
 void hostapd_bss_deinit_no_free(struct hostapd_data *hapd);
 void hostapd_free_hapd_data(struct hostapd_data *hapd);
 void hostapd_cleanup_iface_partial(struct hostapd_iface *iface);
-int hostapd_add_iface(struct hapd_interfaces *ifaces, char *buf);
+enum hostapd_init_rsp hostapd_add_iface(struct hapd_interfaces *ifaces, char *buf);
 int hostapd_remove_iface(struct hapd_interfaces *ifaces, char *buf);
 void hostapd_channel_list_updated(struct hostapd_iface *iface, int initiator);
 void hostapd_set_state(struct hostapd_iface *iface, enum hostapd_iface_state s);
Index: hostapd-2020-06-08-5a8b3662/hostapd/main.c
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/hostapd/main.c
+++ hostapd-2020-06-08-5a8b3662/hostapd/main.c
@@ -859,7 +859,7 @@ int main(int argc, char *argv[])
 		}
 		*fname++ = '\0';
 		iface = hostapd_interface_init_bss(&interfaces, bss_config[i],
-						   fname, debug);
+						   fname, debug, NULL);
 		if (iface == NULL)
 			goto out;
 		for (j = 0; j < interfaces.count; j++) {
Index: hostapd-2020-06-08-5a8b3662/src/ap/hostapd.c
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/src/ap/hostapd.c
+++ hostapd-2020-06-08-5a8b3662/src/ap/hostapd.c
@@ -2610,7 +2610,7 @@ static int ifname_in_use(struct hapd_int
  */
 struct hostapd_iface *
 hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
-			   const char *config_fname, int debug)
+			   const char *config_fname, int debug, enum hostapd_init_rsp *init_rsp)
 {
 	struct hostapd_iface *new_iface = NULL, *iface = NULL;
 	struct hostapd_data *hapd;
@@ -2651,6 +2651,8 @@ hostapd_interface_init_bss(struct hapd_i
 			wpa_printf(MSG_ERROR,
 				   "Interface name %s already in use", ifname);
 			hostapd_config_free(conf);
+			if (init_rsp)
+				*init_rsp = HOSTAPD_INIT_IN_USE;
 			return NULL;
 		}
 
@@ -2993,7 +2995,7 @@ static int hostapd_data_alloc(struct hos
 }
 
 
-int hostapd_add_iface(struct hapd_interfaces *interfaces, char *buf)
+enum hostapd_init_rsp hostapd_add_iface(struct hapd_interfaces *interfaces, char *buf)
 {
 	struct hostapd_config *conf = NULL;
 	struct hostapd_iface *hapd_iface = NULL, *new_iface = NULL;
@@ -3001,22 +3003,23 @@ int hostapd_add_iface(struct hapd_interf
 	char *ptr;
 	size_t i, j;
 	const char *conf_file = NULL, *phy_name = NULL;
+	enum hostapd_init_rsp init_rsp = HOSTAPD_INIT_UNKNOWN;
 
 	if (os_strncmp(buf, "bss_config=", 11) == 0) {
 		char *pos;
 		phy_name = buf + 11;
 		pos = os_strchr(phy_name, ':');
 		if (!pos)
-			return -1;
+			return init_rsp;
 		*pos++ = '\0';
 		conf_file = pos;
 		if (!os_strlen(conf_file))
-			return -1;
+			return init_rsp;
 
 		hapd_iface = hostapd_interface_init_bss(interfaces, phy_name,
-							conf_file, 0);
+							conf_file, 0, &init_rsp);
 		if (!hapd_iface)
-			return -1;
+			return init_rsp;
 		for (j = 0; j < interfaces->count; j++) {
 			if (interfaces->iface[j] == hapd_iface)
 				break;
@@ -3028,7 +3031,7 @@ int hostapd_add_iface(struct hapd_interf
 					       sizeof(struct hostapd_iface *));
 			if (!tmp) {
 				hostapd_interface_deinit_free(hapd_iface);
-				return -1;
+				return init_rsp;
 			}
 			interfaces->iface = tmp;
 			interfaces->iface[interfaces->count++] = hapd_iface;
@@ -3066,16 +3069,16 @@ int hostapd_add_iface(struct hapd_interf
 				hostapd_config_free_bss(hapd->conf);
 				hapd->conf = NULL;
 				os_free(hapd);
-				return -1;
+				return init_rsp;
 			}
 		}
 		hostapd_owe_update_trans(hapd_iface);
-		return 0;
+		return HOSTAPD_INIT_OK;
 	}
 
 	ptr = os_strchr(buf, ' ');
 	if (ptr == NULL)
-		return -1;
+		return init_rsp;
 	*ptr++ = '\0';
 
 	if (os_strncmp(ptr, "config=", 7) == 0)
@@ -3086,7 +3089,7 @@ int hostapd_add_iface(struct hapd_interf
 			       buf)) {
 			wpa_printf(MSG_INFO, "Cannot add interface - it "
 				   "already exists");
-			return -1;
+			return init_rsp;
 		}
 	}
 
@@ -3130,7 +3133,7 @@ int hostapd_add_iface(struct hapd_interf
 	wpa_printf(MSG_INFO, "Add interface '%s'",
 		   hapd_iface->conf->bss[0]->iface);
 
-	return 0;
+	return HOSTAPD_INIT_OK;
 
 fail:
 	if (conf)
@@ -3161,7 +3164,7 @@ fail:
 		}
 		hostapd_cleanup_iface(hapd_iface);
 	}
-	return -1;
+	return init_rsp;
 }
 
 
