--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -1739,7 +1739,7 @@ void handle_probe_req(struct hostapd_dat
 				    csa_offs_len, 0);
 
 	if (ret < 0)
-		wpa_printf(MSG_INFO, "handle_probe_req: send failed");
+		wpa_printf(MSG_DEBUG, "handle_probe_req: send failed");
 
 	os_free(params.resp);
 
--- a/wpa_supplicant/scan.c
+++ b/wpa_supplicant/scan.c
@@ -1947,7 +1947,8 @@ scan:
 	os_free(params.filter_ssids);
 	os_free(params.mac_addr);
 	if (ret) {
-		wpa_msg(wpa_s, MSG_WARNING, "Failed to initiate sched scan");
+		/* Some of the driver implementations do not support scheduled scans */
+		wpa_msg(wpa_s, MSG_DEBUG, "Failed to initiate sched scan");
 		if (prev_state != wpa_s->wpa_state)
 			wpa_supplicant_set_state(wpa_s, prev_state);
 		return ret;
