Index: hostapd-2023-09-08-e5ccbfc6/wpa_supplicant/events.c
===================================================================
--- hostapd-2023-09-08-e5ccbfc6.orig/wpa_supplicant/events.c	2024-04-08 16:17:24.175167674 +0300
+++ hostapd-2023-09-08-e5ccbfc6/wpa_supplicant/events.c	2024-04-08 16:18:35.282734333 +0300
@@ -125,7 +125,7 @@
 {
 	struct wpa_supplicant *wpa_s = eloop_ctx;
 
-	if (wpa_s->disconnected || wpa_s->wpa_state != WPA_SCANNING)
+	if (wpa_s->disconnected)
 		return;
 
 	wpa_dbg(wpa_s, MSG_DEBUG,
