--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -926,7 +926,7 @@ void wpa_supplicant_set_state(struct wpa
 	bool update_fils_connect_params = false;
 #endif /* CONFIG_FILS && IEEE8021X_EAPOL */
 
-	wpa_dbg(wpa_s, MSG_DEBUG, "State: %s -> %s",
+	wpa_dbg(wpa_s, MSG_DEBUG | MSG_LOGFILE, "State: %s -> %s",
 		wpa_supplicant_state_txt(wpa_s->wpa_state),
 		wpa_supplicant_state_txt(state));
 
