--- a/src/ap/vlan_full.c
+++ b/src/ap/vlan_full.c
@@ -755,6 +755,7 @@ full_dynamic_vlan_init(struct hostapd_da
 {
 	struct sockaddr_nl local;
 	struct full_dynamic_vlan *priv;
+	int rcvbuf = TLT_NETLINK_BUFSIZE;
 
 	priv = os_zalloc(sizeof(*priv));
 	if (priv == NULL)
@@ -774,6 +775,11 @@ full_dynamic_vlan_init(struct hostapd_da
 		return NULL;
 	}
 
+	if (setsockopt(priv->s, SOL_SOCKET, SO_RCVBUF, &rcvbuf,
+		   sizeof(rcvbuf)) < 0)
+		wpa_printf(MSG_WARNING, "VLAN: %s: SO_RCVBUF failed: %s",
+			   __func__, strerror(errno));
+
 	os_memset(&local, 0, sizeof(local));
 	local.nl_family = AF_NETLINK;
 	local.nl_groups = RTMGRP_LINK;
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -138,7 +138,7 @@ static void nl80211_register_eloop_read(struct nl_sock **handle,
 	 */
 	int err;
 
-	err = nl_socket_set_buffer_size(*handle, 262144, 0);
+	err = nl_socket_set_buffer_size(*handle, TLT_NL80211_BUFSIZE, 0);
 	if (err < 0) {
 		wpa_printf(MSG_DEBUG,
 			   "nl80211: Could not set nl_socket RX buffer size: %s",
--- a/src/drivers/netlink.c
+++ b/src/drivers/netlink.c
@@ -92,6 +92,7 @@ struct netlink_data * netlink_init(struc
 {
 	struct netlink_data *netlink;
 	struct sockaddr_nl local;
+	int rcvbuf = TLT_NETLINK_BUFSIZE;
 
 	netlink = os_zalloc(sizeof(*netlink));
 	if (netlink == NULL)
@@ -105,6 +106,11 @@ struct netlink_data * netlink_init(struc
 		return NULL;
 	}
 
+	if (setsockopt(netlink->sock, SOL_SOCKET, SO_RCVBUF, &rcvbuf,
+		   sizeof(rcvbuf)) < 0)
+		wpa_printf(MSG_WARNING, "netlink: SO_RCVBUF failed: %s",
+			   strerror(errno));
+
 	os_memset(&local, 0, sizeof(local));
 	local.nl_family = AF_NETLINK;
 	local.nl_groups = RTMGRP_LINK;
