Index: xtables-addons-3.13/extensions/xt_ECHO.c
===================================================================
--- xtables-addons-3.13.orig/extensions/xt_ECHO.c
+++ xtables-addons-3.13/extensions/xt_ECHO.c
@@ -97,7 +97,11 @@ echo_tg6(struct sk_buff *oldskb, const s
 	memcpy(&fl.daddr, &newip->daddr, sizeof(fl.daddr));
 	fl.fl6_sport = newudp->source;
 	fl.fl6_dport = newudp->dest;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0)
 	security_skb_classify_flow((struct sk_buff *)oldskb, flowi6_to_flowi(&fl));
+#else
+	security_skb_classify_flow((struct sk_buff *)oldskb, flowi6_to_flowi_common(&fl));
+#endif
 	dst = ip6_route_output(net, NULL, &fl);
 	if (dst == NULL || dst->error != 0) {
 		dst_release(dst);
