Index: xtables-addons-3.24/extensions/compat_xtables.h
===================================================================
--- xtables-addons-3.24.orig/extensions/compat_xtables.h
+++ xtables-addons-3.24/extensions/compat_xtables.h
@@ -21,7 +21,7 @@
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0) || \
     LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 9) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0) || \
     LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 78) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0) || \
-    LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 158) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 20, 0)
+    LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 20, 0)
 #else
 #	define ip_route_me_harder(xnet, xsk, xskb, xaddrtype) ip_route_me_harder((xnet), (xskb), (xaddrtype))
 #	define ip6_route_me_harder(xnet, xsk, xskb) ip6_route_me_harder((xnet), (xskb))
Index: xtables-addons-3.24/extensions/xt_DELUDE.c
===================================================================
--- xtables-addons-3.24.orig/extensions/xt_DELUDE.c
+++ xtables-addons-3.24/extensions/xt_DELUDE.c
@@ -123,7 +123,11 @@ static void delude_send_reset(struct sk_
 
 	/* ip_route_me_harder expects skb->dst to be set */
 	skb_dst_set(nskb, dst_clone(skb_dst(oldskb)));
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
+	if (ip_route_me_harder(net, nskb, addr_type))
+#else
 	if (ip_route_me_harder(par_net(par), par->state->sk, nskb, addr_type))
+#endif
 		goto free_nskb;
 	else
 		niph = ip_hdr(nskb);
Index: xtables-addons-3.24/extensions/xt_ECHO.c
===================================================================
--- xtables-addons-3.24.orig/extensions/xt_ECHO.c
+++ xtables-addons-3.24/extensions/xt_ECHO.c
@@ -200,8 +200,11 @@ echo_tg4(struct sk_buff *oldskb, const s
 
 	/* ip_route_me_harder expects the skb's dst to be set */
 	skb_dst_set(newskb, dst_clone(skb_dst(oldskb)));
-	if (ip_route_me_harder(par_net(par), par->state->sk, newskb,
-	    RTN_UNSPEC) != 0)
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
+	if (ip_route_me_harder(par_net(par), newskb, RTN_UNSPEC) != 0)
+#else
+	if (ip_route_me_harder(par_net(par), par->state->sk, newskb, RTN_UNSPEC) != 0)
+#endif
 		goto free_nskb;
 
 	newip->ttl = ip4_dst_hoplimit(skb_dst(newskb));
Index: xtables-addons-3.24/extensions/xt_TARPIT.c
===================================================================
--- xtables-addons-3.24.orig/extensions/xt_TARPIT.c
+++ xtables-addons-3.24/extensions/xt_TARPIT.c
@@ -267,7 +267,11 @@ static void tarpit_tcp4(const struct xt_
 #endif
 		addr_type = RTN_LOCAL;
 
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
+	if (ip_route_me_harder(net, nskb, addr_type))
+#else
 	if (ip_route_me_harder(par_net(par), par->state->sk, nskb, addr_type) != 0)
+#endif
 		goto free_nskb;
 	else
 		niph = ip_hdr(nskb);
@@ -400,7 +404,11 @@ static void tarpit_tcp6(const struct xt_
 	              &ipv6_hdr(nskb)->daddr, sizeof(struct tcphdr),
 	              IPPROTO_TCP,
 	              csum_partial(tcph, sizeof(struct tcphdr), 0));
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
+	if (ip6_route_me_harder(net, nskb))
+#else
 	if (ip6_route_me_harder(par_net(par), nskb->sk, nskb))
+#endif
 		goto free_nskb;
 
 	nskb->ip_summed = CHECKSUM_NONE;
