Index: net-snmp-5.9.5/configure
===================================================================
--- net-snmp-5.9.5.orig/configure
+++ net-snmp-5.9.5/configure
@@ -25008,7 +25008,7 @@ while test "x$new_transport_list" != "x"
       if test -f "$srcdir/$rel_transport_src"; then
         transport_result_list="$transport_result_list $i"
         transport_src_list="$transport_src $transport_src_list"
-        if test -f "$srcdir/$rel_transport_hdr"; then
+        if test -f "$srcdir/$rel_transport_hdr" && test "x$i" != "xIPBase"; then
           transport_hdr_list="$transport_hdr $transport_hdr_list"
         fi
         transport_def=`echo $i | $SED 'y/abcdefghijklmnopqrstuvwxyz/ABCDEFGHIJKLMNOPQRSTUVWXYZ/'`
Index: net-snmp-5.9.5/configure.d/config_modules_transports
===================================================================
--- net-snmp-5.9.5.orig/configure.d/config_modules_transports
+++ net-snmp-5.9.5/configure.d/config_modules_transports
@@ -60,7 +60,7 @@ while test "x$new_transport_list" != "x"
       if test -f "$srcdir/$rel_transport_src"; then
         transport_result_list="$transport_result_list $i"
         transport_src_list="$transport_src $transport_src_list"
-        if test -f "$srcdir/$rel_transport_hdr"; then
+        if test -f "$srcdir/$rel_transport_hdr" && test "x$i" != "xIPBase"; then
           transport_hdr_list="$transport_hdr $transport_hdr_list"
         fi
         transport_def=`echo $i | $SED 'y/abcdefghijklmnopqrstuvwxyz/ABCDEFGHIJKLMNOPQRSTUVWXYZ/'`
