--- a/platform/openwrt/sqm-hotplug
+++ b/platform/openwrt/sqm-hotplug
@@ -1,11 +1,12 @@
 [ -n "$DEVICE" ] || exit 0
 
+SQM_RUN="/usr/local/usr/lib/sqm/run.sh"
 ALL_DEVICES=$(echo $DEVICE $(uci -q get network.$INTERFACE.ifname) | tr ' ' '\n' | sort -u)
 
 restart_sqm() {
     for dev in $ALL_DEVICES; do
-        /usr/lib/sqm/run.sh stop $dev
-        /usr/lib/sqm/run.sh start $dev
+        $SQM_RUN stop $dev
+        $SQM_RUN start $dev
     done
 }
 
@@ -13,6 +14,6 @@ restart_sqm() {
 
 if [ "$ACTION" = ifdown ]; then
     for dev in $ALL_DEVICES; do
-        /usr/lib/sqm/run.sh stop $dev
+        $SQM_RUN stop $dev
     done
 fi
--- a/platform/openwrt/sqm-init
+++ b/platform/openwrt/sqm-init
@@ -2,6 +2,7 @@
 
 START=50
 USE_PROCD=1
+SQM_RUN="/usr/local/usr/lib/sqm/run.sh"
 
 service_triggers()
 {
@@ -16,12 +17,12 @@ reload_service()
 
 start_service()
 {
-	/usr/lib/sqm/run.sh start "$@"
+	$SQM_RUN start "$@"
 }
 
 stop_service()
 {
-	/usr/lib/sqm/run.sh stop "$@"
+	$SQM_RUN stop "$@"
 }
 
 boot()
--- a/platform/openwrt/sqm.conf
+++ b/platform/openwrt/sqm.conf
@@ -1,4 +1,4 @@
-SQM_LIB_DIR=/usr/lib/sqm
+SQM_LIB_DIR=/usr/local/usr/lib/sqm
 SQM_STATE_DIR=/var/run/sqm
 SQM_QDISC_STATE_DIR=${SQM_STATE_DIR}/available_qdiscs
 SQM_CHECK_QDISCS="fq_codel codel pie sfq cake"
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,5 @@
 PREFIX:=/usr
+USRDIR:=
 DESTDIR:=
 PLATFORM:=linux
 PKG_CONFIG:=$(shell which pkg-config 2>/dev/null)
@@ -18,7 +19,7 @@ install-openwrt: install-lib
 		$(DESTDIR)/etc/init.d
 	install -m 0600 platform/openwrt/sqm-hotplug $(DESTDIR)/etc/hotplug.d/iface/11-sqm
 	install -m 0755 platform/openwrt/sqm-init $(DESTDIR)/etc/init.d/sqm
-	install -m 0744 src/run-openwrt.sh $(DESTDIR)$(PREFIX)/lib/sqm/run.sh
+	install -m 0744 src/run-openwrt.sh $(USRDIR)$(PREFIX)/lib/sqm/run.sh
 
 .PHONY: install-linux
 install-linux: install-lib
@@ -34,12 +35,12 @@ install-linux: install-lib
 
 .PHONY: install-lib
 install-lib:
-	install -m 0755 -d $(DESTDIR)/etc/sqm $(DESTDIR)$(PREFIX)/lib/sqm
+	install -m 0755 -d $(DESTDIR)/etc/sqm $(USRDIR)$(PREFIX)/lib/sqm
 	install -m 0644 -C -b platform/$(PLATFORM)/sqm.conf $(DESTDIR)/etc/sqm/sqm.conf
 	install -m 0644  src/functions.sh src/defaults.sh \
-		src/*.qos src/*.help $(DESTDIR)$(PREFIX)/lib/sqm
+		src/*.qos src/*.help $(USRDIR)$(PREFIX)/lib/sqm
 	install -m 0744  src/start-sqm src/stop-sqm src/update-available-qdiscs \
-		$(DESTDIR)$(PREFIX)/lib/sqm
+		$(USRDIR)$(PREFIX)/lib/sqm
 
 .PHONY: uninstall
 uninstall: uninstall-$(PLATFORM)
@@ -59,6 +60,6 @@ uninstall-linux: uninstall-lib
 
 .PHONY: uninstall-lib
 uninstall-lib:
-	@for f in $(DESTDIR)$(PREFIX)/lib/sqm/*; do \
+	@for f in $(USRDIR)$(PREFIX)/lib/sqm/*; do \
 		if [ -f "$$f" ]; then rm -vf "$$f"; fi; done
-	@rmdir -v $(DESTDIR)$(PREFIX)/lib/sqm
+	@rmdir -v $(USRDIR)$(PREFIX)/lib/sqm
