Index: libiwinfo-2023-05-17-c9f5c3f7/iwinfo_nl80211.c
===================================================================
--- libiwinfo-2023-05-17-c9f5c3f7.orig/iwinfo_nl80211.c
+++ libiwinfo-2023-05-17-c9f5c3f7/iwinfo_nl80211.c
@@ -845,7 +845,7 @@ static char * nl80211_phy2ifname(const c
 			clmode = nl80211_readint(buffer);
 
 			/* prefer non-supplicant-based devices */
-			if ((ifidx < 0) || (cifidx < ifidx) || ((lmode == 1) && (clmode != 1)))
+			if ((ifidx < 0) || (cifidx > ifidx) || ((lmode == 1) && (clmode != 1)))
 			{
 				ifidx = cifidx;
 				lmode = clmode;
