Index: libiwinfo-2022-08-19-0dad3e66/iwinfo_nl80211.c
===================================================================
--- libiwinfo-2022-08-19-0dad3e66.orig/iwinfo_nl80211.c	2023-10-20 09:23:54.859589251 +0300
+++ libiwinfo-2022-08-19-0dad3e66/iwinfo_nl80211.c	2023-10-20 09:25:58.366762966 +0300
@@ -741,7 +741,7 @@
 					         "/sys/class/net/%s/ifindex", e->d_name);
 
 					if ((cifidx = nl80211_readint(buffer)) >= 0 &&
-					    ((ifidx < 0) || (cifidx < ifidx)))
+					    ((ifidx < 0) || (cifidx > ifidx)))
 					{
 						ifidx = cifidx;
 						strncpy(nif, e->d_name, sizeof(nif) - 1);
