--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -2716,9 +2716,16 @@ static int nl80211_get_scanlist_cb(struc
 static int nl80211_get_scanlist_nl(const char *ifname, char *buf, int *len)
 {
 	struct nl80211_scanlist sl = { .e = (struct iwinfo_scanlist_entry *)buf };
+	int ret, tries = 0;
 
-	if (nl80211_request(ifname, NL80211_CMD_TRIGGER_SCAN, 0, NULL, NULL))
-		goto out;
+	while ((ret = nl80211_request(ifname, NL80211_CMD_TRIGGER_SCAN, 0, NULL, NULL)))
+	{
+		if (tries >= 5 || ret != -EBUSY)
+			goto out;
+
+		tries++;
+		sleep(1);
+	}
 
 	if (nl80211_wait("nl80211", "scan",
 	                 NL80211_CMD_NEW_SCAN_RESULTS, NL80211_CMD_SCAN_ABORTED))
