--- a/rcS.c
+++ b/rcS.c
@@ -33,6 +33,7 @@
 
 #include "procd.h"
 #include "rcS.h"
+#include "utils/utils.h"
 
 static struct runqueue q, r;
 static LIST_HEAD(services);
@@ -112,6 +113,8 @@ static void q_initd_run(struct runqueue
 	if (devnull > STDERR_FILENO)
 		close(devnull);
 
+	patch_stdio("/dev/console");
+
 	execlp(s->file, s->file, s->param, NULL);
 	exit(1);
 }
--- a/initd/preinit.c
+++ b/initd/preinit.c
@@ -29,6 +29,7 @@
 #include "init.h"
 #include "../watchdog.h"
 #include "../sysupgrade.h"
+#include "../utils/utils.h"
 
 static struct uloop_process preinit_proc;
 static struct uloop_process plugd_proc;
@@ -132,6 +133,7 @@ preinit(void)
 	plugd_proc.cb = plugd_proc_cb;
 	plugd_proc.pid = fork();
 	if (!plugd_proc.pid) {
+		patch_stdio("/dev/kmsg");
 		execvp(plug[0], plug);
 		ERROR("Failed to start plugd: %m\n");
 		exit(EXIT_FAILURE);
@@ -154,6 +156,7 @@ preinit(void)
 	preinit_proc.cb = spawn_procd;
 	preinit_proc.pid = fork();
 	if (!preinit_proc.pid) {
+		patch_stdio("/dev/kmsg");
 		execvp(init[0], init);
 		ERROR("Failed to start preinit: %m\n");
 		exit(EXIT_FAILURE);
