--- a/initd/preinit.c     2021-02-23 02:42:37.000000000 +0200
+++ b/initd/preinit.c     2023-12-05 11:03:45.825608127 +0200
@@ -29,6 +29,7 @@
 #include "init.h"
 #include "../watchdog.h"
 #include "../sysupgrade.h"
+#include "../utils/utils.h"
 
 static struct uloop_process preinit_proc;
 static struct uloop_process plugd_proc;
@@ -132,6 +133,7 @@
	plugd_proc.cb = plugd_proc_cb;
	plugd_proc.pid = fork();
	if (!plugd_proc.pid) {
+		patch_stdio("/dev/kmsg");
		execvp(plug[0], plug);
		ERROR("Failed to start plugd: %m\n");
		exit(EXIT_FAILURE);
@@ -154,6 +156,7 @@
	preinit_proc.cb = spawn_procd;
	preinit_proc.pid = fork();
	if (!preinit_proc.pid) {
+		patch_stdio("/dev/kmsg");
		execvp(init[0], init);
		ERROR("Failed to start preinit: %m\n");
		exit(EXIT_FAILURE);

