Index: mwan3-2.10.12/files/etc/hotplug.d/iface/15-mwan3
===================================================================
--- mwan3-2.10.12.orig/files/etc/hotplug.d/iface/15-mwan3
+++ mwan3-2.10.12/files/etc/hotplug.d/iface/15-mwan3
@@ -79,6 +79,9 @@ case "$ACTION" in
 		[ "$ACTION" = ifup ] && procd_running mwan3 "track_$INTERFACE" && procd_send_signal mwan3 "track_$INTERFACE" USR2
 		;;
 	disconnected)
+		iface_disabled=$(uci_get network ${INTERFACE} disabled)
+		[ "$iface_disabled" -eq 1 ] && echo "disabled" > /var/run/mwan3track/${INTERFACE}/STATUS
+
 		echo "disconnected" > /var/run/mwan3track/${INTERFACE}/STATE
 		mwan3_set_iface_hotplug_state $INTERFACE "offline"
 		mwan3_set_policies_iptables
Index: mwan3-2.10.12/files/usr/sbin/mwan3track
===================================================================
--- mwan3-2.10.12.orig/files/usr/sbin/mwan3track
+++ mwan3-2.10.12/files/usr/sbin/mwan3track
@@ -416,8 +416,8 @@ main() {
 
 		if [ "${IFDOWN_EVENT}" -eq 1 ]; then
 			LOG debug "Register ifdown event on interface ${INTERFACE} (${DEVICE})"
-			disabled
 			disconnected
+			disabled
 			IFDOWN_EVENT=0
 		fi
 		if [ "${IFUP_EVENT}" -eq 1 ]; then
