Index: mwan3-2.10.12/files/usr/share/15-mwan3
===================================================================
--- mwan3-2.10.12.orig/files/usr/share/15-mwan3
+++ mwan3-2.10.12/files/usr/share/15-mwan3
@@ -115,8 +115,34 @@ flush_conntrack() {
 	ipsec down "${conn_name}-${1}"
 }
 
+restart_openvpn() {
+	local sid="$1"
+	local enable inst_type active
+	local restart=0
+	config_get enable "$sid" enable
+	config_get inst_type "$sid" type
+	active="$(cat /tmp/run/mwan3/active_wan)"
+	if [ "$MWAN3_STARTUP" != "init" ] && [ "$enable" = "1" ] && [ "$inst_type" = "client" ]; then
+		case "$ACTION" in
+			ifup|connected)
+				[ "$IFNAME" = "$active" ] && restart=1
+				;;
+			ifdown|disconnected)
+				[ "$IFNAME" != "$active" ] && restart=1
+				;;
+		esac
+		if [ "$restart" = 1 ]; then
+			{ sleep 3
+			logger -t mwan3 "OpenVPN interface $sid is restarting"
+			/etc/init.d/openvpn restart "$sid"; } &
+		fi
+	fi
+}
+
 config_load ipsec
 config_foreach flush_conntrack connection
+config_load openvpn
+config_foreach restart_openvpn openvpn
 config_load mwan3
 
 exit 0
