--- a/files/lib/mwan3/mwan3.sh
+++ b/files/lib/mwan3/mwan3.sh
@@ -481,6 +481,29 @@ mwan3_delete_iface_route()
 	fi
 }
 
+mwan3_create_wan_source_rule()
+{
+	local id family IP
+
+	config_get family "$1" family ipv4
+	mwan3_get_iface_id id "$1"
+
+	[ -n "$id" ] || return 0
+
+	if [ "$family" = "ipv4" ]; then
+		IP="$IP4"
+		network_get_ipaddr wan_addr "${1}"
+	elif [ "$family" = "ipv6" ] && [ $NO_IPV6 -eq 0 ]; then
+		IP="$IP6"
+		network_get_subnet6 wan_addr "${1}"
+	else
+		return
+	fi
+
+	mwan3_delete_wan_source_rule "$1"
+	[ -n "$wan_addr" ] && $IP rule add pref $((id+1500)) from ${wan_addr} lookup "$id"
+}
+
 mwan3_create_iface_rules()
 {
 	local id family IP
@@ -500,11 +523,31 @@ mwan3_create_iface_rules()
 
 	mwan3_delete_iface_rules "$1"
 
+	mwan3_create_wan_source_rule "$1"
 	$IP rule add pref $((id+1000)) iif "$2" lookup "$id"
 	$IP rule add pref $((id+2000)) fwmark "$(mwan3_id2mask id MMX_MASK)/$MMX_MASK" lookup "$id"
 	$IP rule add pref $((id+3000)) fwmark "$(mwan3_id2mask id MMX_MASK)/$MMX_MASK" unreachable
 }
 
+mwan3_delete_wan_source_rule()
+{
+	local id family IP
+	config_get family "$1" family ipv4
+	mwan3_get_iface_id id "$1"
+
+	[ -n "$id" ] || return 0
+
+	if [ "$family" = "ipv4" ]; then
+		IP="$IP4"
+	elif [ "$family" = "ipv6" ] && [ $NO_IPV6 -eq 0 ]; then
+		IP="$IP6"
+	else
+		return
+	fi
+
+	$IP rule del prio $((id+1500)) 2>/dev/null
+}
+
 mwan3_delete_iface_rules()
 {
 	local id family IP rule_id
@@ -522,7 +565,7 @@ mwan3_delete_iface_rules()
 		return
 	fi
 
-	for rule_id in $(ip rule list | awk '$1 % 1000 == '$id' && $1 > 1000 && $1 < 4000 {print substr($1,0,4)}'); do
+	for rule_id in $(ip rule list | awk '($1 % 1000 == '$id' || $1 % 1500 == '$id') && $1 > 1000 && $1 < 4000 {print substr($1,0,4)}'); do
 		$IP rule del pref $rule_id
 	done
 }
--- a/files/usr/share/15-mwan3
+++ b/files/usr/share/15-mwan3
@@ -71,6 +71,7 @@ case "$ACTION" in
 		mwan3_set_iface_hotplug_state $IFNAME "online"
 		mwan3_set_policies_iptables
 		mwan3_set_event $IFNAME
+		mwan3_create_wan_source_rule $IFNAME
 		;;
 	ifup)
 		mwan3_get_true_iface true_ifname $IFNAME
@@ -98,6 +99,7 @@ case "$ACTION" in
 		mwan3_set_iface_hotplug_state $IFNAME "offline"
 		mwan3_set_policies_iptables
 		mwan3_set_fail_event $IFNAME
+		mwan3_delete_wan_source_rule $IFNAME
 		;;
 	ifdown)
 		mwan3_set_iface_hotplug_state $IFNAME "offline"
