index 1270d63c8..e10d6ca03 100644
--- a/src/server/ua_services_discovery.c
+++ b/src/server/ua_services_discovery.c
@@ -294,9 +294,9 @@ getDefaultEncryptedSecurityPolicy(UA_Server *server) {
         if(!UA_String_equal(&UA_SECURITY_POLICY_NONE_URI, &sp->policyUri))
             return sp;
     }
-    UA_LOG_WARNING(server->config.logging, UA_LOGCATEGORY_CLIENT,
-                   "Could not find a SecurityPolicy with encryption for the "
-                   "UserTokenPolicy. Using an unencrypted policy.");
+    // UA_LOG_WARNING(server->config.logging, UA_LOGCATEGORY_CLIENT,
+    //                "Could not find a SecurityPolicy with encryption for the "
+    //                "UserTokenPolicy. Using an unencrypted policy.");
     return server->config.securityPoliciesSize > 0 ?
         &server->config.securityPolicies[0]: NULL;
 }
