Index: ovpn-dco-0.2.20241216/drivers/net/ovpn-dco/ovpn.c
===================================================================
--- ovpn-dco-0.2.20241216.orig/drivers/net/ovpn-dco/ovpn.c
+++ ovpn-dco-0.2.20241216/drivers/net/ovpn-dco/ovpn.c
@@ -22,7 +22,9 @@
 #include "udp.h"
 
 #include <linux/workqueue.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,5,0)
 #include <net/gso.h>
+#endif
 #include <uapi/linux/if_ether.h>
 
 static const unsigned char ovpn_keepalive_message[] = {
