--- a/lua/uci.c
+++ b/lua/uci.c
@@ -702,7 +702,6 @@ static int
 uci_lua_package_cmd(lua_State *L, enum pkg_cmd cmd)
 {
 	struct uci_context *ctx;
-	struct uci_element *e, *tmp;
 	struct uci_ptr ptr;
 	char *s = NULL;
 	int nargs, offset = 0;
@@ -717,24 +716,20 @@ uci_lua_package_cmd(lua_State *L, enum pkg_cmd cmd)

 	lookup_ptr(ctx, &ptr, NULL, true);

-	uci_foreach_element_safe(&ctx->root, tmp, e) {
-		struct uci_package *p = uci_to_package(e);
-
-		if (ptr.p && (ptr.p != p))
-			continue;
+	if (!ptr.p) {
+		goto err;
+	}

-		ptr.p = p;
-		switch(cmd) {
-		case CMD_COMMIT:
-			uci_commit(ctx, &p, false);
-			break;
-		case CMD_SAVE:
-			uci_save(ctx, p);
-			break;
-		case CMD_REVERT:
-			uci_revert(ctx, &ptr);
-			break;
-		}
+	switch(cmd) {
+	case CMD_COMMIT:
+		uci_commit(ctx, &ptr.p, false);
+		break;
+	case CMD_SAVE:
+		uci_save(ctx, ptr.p);
+		break;
+	case CMD_REVERT:
+		uci_revert(ctx, &ptr);
+		break;
 	}

 err:
