From 96668b39a274c2ec1c212c7b5112f0e6bde1adf1 Mon Sep 17 00:00:00 2001
From: Joris Vaisvila <joris.vaisvila@teltonika.lt>
Date: Mon, 27 Jan 2025 14:56:49 +0200
Subject: [PATCH 1/1] uci: chown with silent fail

---
 file.c | 1 +
 1 file changed, 1 insertion(+)

Index: uci-2021-10-22-f84f49f0-22/file.c
===================================================================
--- uci-2021-10-22-f84f49f0-22.orig/file.c
+++ uci-2021-10-22-f84f49f0-22/file.c
@@ -821,10 +821,13 @@ done:
 	uci_close_stream(f1);
 	if (do_rename) {
 		path = realpath(p->path, NULL);
-		if (!path || stat(path, &statbuf) || chmod(filename, statbuf.st_mode) || rename(filename, path)) {
+		if (!path || stat(path, &statbuf) || chmod(filename, statbuf.st_mode) ||
+				chown(filename, getuid() ? (uid_t)-1 : statbuf.st_uid, statbuf.st_gid) ||
+				rename(filename, path)) {
 			unlink(filename);
 			UCI_THROW(ctx, UCI_ERR_IO);
 		}
+
 		free(path);
 	}
 	if (ctx->err)
