Index: tailscale-1.86.2/util/linuxfw/iptables_runner.go
===================================================================
--- tailscale-1.86.2.orig/util/linuxfw/iptables_runner.go
+++ tailscale-1.86.2/util/linuxfw/iptables_runner.go
@@ -88,7 +88,7 @@ func newIPTablesRunner(logf logger.Logf)
 			return nil, err
 		}
 		supportsV6Filter = checkSupportsV6Filter(ipt6, logf)
-		supportsV6NAT = checkSupportsV6NAT(ipt6, logf)
+		// supportsV6NAT = checkSupportsV6NAT(ipt6, logf)
 		logf("netfilter running in iptables mode v6 = %v, v6filter = %v, v6nat = %v", supportsV6, supportsV6Filter, supportsV6NAT)
 	}
 	return &iptablesRunner{
