Index: tailscale-1.70.0/cmd/tailscaled/tailscaled.go
===================================================================
--- tailscale-1.70.0.orig/cmd/tailscaled/tailscaled.go
+++ tailscale-1.70.0/cmd/tailscaled/tailscaled.go
@@ -67,6 +67,7 @@ import (
 	"tailscale.com/wgengine"
 	"tailscale.com/wgengine/netstack"
 	"tailscale.com/wgengine/router"
+	"kernel.org/pub/linux/libs/security/libcap/cap"
 )
 
 // defaultTunName returns the default tun device name for the platform.
@@ -151,6 +152,23 @@ var subCommands = map[string]*func([]str
 
 var beCLI func() // non-nil if CLI is linked in
 
+func setInheritCapabilities() error {
+	caps := cap.GetProc()
+
+	if err := caps.SetFlag(cap.Inheritable, true, cap.NET_ADMIN); err != nil {
+		return err
+	}
+	if err := caps.SetFlag(cap.Inheritable, true, cap.NET_RAW); err != nil {
+		return err
+	}
+
+	if err := caps.SetProc(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
 func main() {
 	envknob.PanicIfAnyEnvCheckedInInit()
 	envknob.ApplyDiskConfig()
@@ -171,6 +189,10 @@ func main() {
 	flag.BoolVar(&args.disableLogs, "no-logs-no-support", false, "disable log uploads; this also disables any technical support")
 	flag.StringVar(&args.confFile, "config", "", "path to config file, or 'vm:user-data' to use the VM's user-data (EC2)")
 
+	if err := setInheritCapabilities(); err != nil {
+		log.Fatal("Failed to set capabilities: %v", err)
+	}
+
 	if len(os.Args) > 0 && filepath.Base(os.Args[0]) == "tailscale" && beCLI != nil {
 		beCLI()
 		return
Index: tailscale-1.70.0/go.mod
===================================================================
--- tailscale-1.70.0.orig/go.mod
+++ tailscale-1.70.0/go.mod
@@ -390,4 +390,6 @@ require (
 	mvdan.cc/unparam v0.0.0-20230312165513-e84e2d14e3b8 // indirect
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
+	kernel.org/pub/linux/libs/security/libcap/cap v1.2.76 // indirect
+	kernel.org/pub/linux/libs/security/libcap/psx v1.2.76 // indirect
 )
Index: tailscale-1.70.0/go.sum
===================================================================
--- tailscale-1.70.0.orig/go.sum
+++ tailscale-1.70.0/go.sum
@@ -1513,6 +1513,10 @@ mvdan.cc/unparam v0.0.0-20230312165513-e
 mvdan.cc/unparam v0.0.0-20230312165513-e84e2d14e3b8/go.mod h1:Oh/d7dEtzsNHGOq1Cdv8aMm3KdKhVvPbRQcM8WFpBR8=
 nhooyr.io/websocket v1.8.10 h1:mv4p+MnGrLDcPlBoWsvPP7XCzTYMXP9F9eIGoKbgx7Q=
 nhooyr.io/websocket v1.8.10/go.mod h1:rN9OFWIUwuxg4fR5tELlYC04bXYowCP9GX47ivo2l+c=
+kernel.org/pub/linux/libs/security/libcap/cap v1.2.76 h1:mrdLPj8ujM6eIKGtd1PkkuCIodpFFDM42Cfm0YODkIM=
+kernel.org/pub/linux/libs/security/libcap/cap v1.2.76/go.mod h1:7V2BQeHnVAQwhCnCPJ977giCeGDiywVewWF+8vkpPlc=
+kernel.org/pub/linux/libs/security/libcap/psx v1.2.76 h1:3DyzQ30OHt3wiOZVL1se2g1PAPJIU7+tMUyvfMUj1dY=
+kernel.org/pub/linux/libs/security/libcap/psx v1.2.76/go.mod h1:+l6Ee2F59XiJ2I6WR5ObpC1utCQJZ/VLsEbQCD8RG24=
 rsc.io/binaryregexp v0.2.0/go.mod h1:qTv7/COck+e2FymRvadv62gMdZztPaShugOCi3I+8D8=
 rsc.io/quote/v3 v3.1.0/go.mod h1:yEA65RcK8LyAZtP9Kv3t0HmxON59tX3rD+tICJqUlj0=
 rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9JXDnKaTXpA=
